<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.19.1 by Michael Rose
  Copyright 2013-2019 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>Actinomyces comparative genomics - Daniel Utter</title>
<meta name="description" content="From Utter et al., 2020 ISME ">


  <meta name="author" content="du">


<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Daniel Utter">
<meta property="og:title" content="Actinomyces comparative genomics">
<meta property="og:url" content="http://localhost:4000/projects/Act23_phylo">


  <meta property="og:description" content="From Utter et al., 2020 ISME ">







  <meta property="article:published_time" content="2020-10-16T21:53:26-04:00">






<link rel="canonical" href="http://localhost:4000/projects/Act23_phylo">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": null,
      "url": "http://localhost:4000/"
    
  }
</script>






<!-- end _includes/seo.html -->


<link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Daniel Utter Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">

<!--[if IE]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--single">
    <nav class="skip-links">
  <h2 class="screen-reader-text">Skip links</h2>
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          Daniel Utter
          
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/About/about.md">About</a>
            </li><li class="masthead__menu-item">
              <a href="/CV/CV.md">CV</a>
            </li><li class="masthead__menu-item">
              <a href="/Research/Research.md">Research</a>
            </li><li class="masthead__menu-item">
              <a href="/Methods/methods.md">Reproducible Methods</a>
            </li><li class="masthead__menu-item">
              <a href="/Misc/omics.md">Essential 'omics commands</a>
            </li></ul>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      



<div id="main" role="main">
  


  <article class="page" itemscope itemtype="https://schema.org/CreativeWork">
    <meta itemprop="headline" content="Actinomyces comparative genomics">
    <meta itemprop="description" content="From Utter et al., 2020 ISME">
    <meta itemprop="datePublished" content="2020-10-16T21:53:26-04:00">
    

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline"><em>Actinomyces</em> comparative genomics
</h1>
          
        </header>
      

      <section class="page__content" itemprop="text">
        
          <aside class="sidebar__right sticky">
            <nav class="toc">
              <header><h4 class="nav__title"><i class="fas fa-file-alt"></i> On this page</h4></header>
              <ul class="toc__menu">
  <li><a href="#overview">Overview</a></li>
  <li><a href="#pangenome-construction">Pangenome construction</a></li>
  <li><a href="#annotating-gene-calls">Annotating gene calls</a></li>
  <li><a href="#adding-some-miscellaneous-data">Adding some miscellaneous data</a></li>
  <li><a href="#investigating-single-gene-trees-from-core-gene-clusters">Investigating single-gene trees from core gene clusters</a></li>
  <li><a href="#extracting-genes-and-generating-gene-trees">Extracting genes and generating gene trees</a></li>
  <li><a href="#identifying-gene-clusters-producing-interesting-topologies">Identifying gene clusters producing interesting topologies</a></li>
  <li><a href="#phylogenetic-analysis-1-amino-acid-identity-aai">Phylogenetic analysis 1: Amino Acid Identity (AAI)</a></li>
  <li><a href="#phylogenetic-analysis-2-phylophlan2">Phylogenetic analysis 2: PhyloPhlAn2</a></li>
  <li><a href="#comparing-phylophlan-with-the-pangenome-dendrogram">Comparing phylophlan with the pangenome dendrogram</a></li>
</ul>

            </nav>
          </aside>
        
        <h2 id="overview">Overview</h2>

<p>This is a narrative methods providing a reproducible workflow of the
analyses used in our paper <strong>bioRxiv link</strong>. In this paper, we describe
how we found that <em>Nanosynbacter lyticus</em> TM7x can establish an
ectosymbiotic relationship on several related <em>Actinomyces</em> spp., but
the inital relationship differs among the susceptible strains. That is,
TM7x behaves like a parasite with <em>A. odontolyticus</em> XH001 in the lab,
<a href="https://www.pnas.org/content/115/48/12277">documented here</a>, where
naive XH001 experience a strong crash in growth upon exposure to TM7x,
and after a few passages they regain balance and achieve a stable
relationship with TM7x. However, not all <em>Actinomyces</em> strains that
could host TM7x had this crash. Specifically, out of 13 susceptible
<em>Actinomyces</em> strains spanning ~74% AAI and two named species, three
strains never crashed in response to TM7x, regardless of the initial
TM7x dosage. This observation is exciting, as it suggests that across a
moderate host range, the nature of the symbiosis, e.g., parasitism,
might not be the same across all potential hosts. But further, since we
have genomes for the majority of these strains, we might glean some some
insights into what makes <em>A</em>. spp. able to host TM7x or why some strains
are negatively vs. neutrally affected by hosting TM7x.</p>

<p>First we manually downloaded genomes from NCBI corresponding to the 23
strains we characterized experimentally. For convenience here are their
names:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>A_odontolyticus_ATCC_17929
A_odontolyticus_ATCC_17982
A_meyeri_ATCC_35568
A_odontolyticus_F0309
A_sp_HMT_180_F0310
A_sp_HMT_172_F0311
A_sp_HMT_849_F0330
A_sp_HMT_848_F0332
A_sp_HMT_171_F0337
A_sp_HMT_178_F0338
A_sp_HMT_175_F0384
A_sp_HMT_170_F0386
A_sp_HMT_448_F0400
A_massiliensis_F0489
A_johnsonii_F0510
A_graevenitzii_F0530
A_sp_HMT_877_F0543
A_sp_HMT_414_F0588
A_sp_ICM39
A_sp_ICM47
A_sp_ICM58
A_meyeri_W712
A_odontolyticus_XH001
</code></pre></div></div>

<p>While collecting the genomes, we changed each fasta’s defline to be like
<code class="highlighter-rouge">&gt;A_meyeri_W712_ctg1</code> from the more complicated default NCBI header.
There are many ways to do this but we did it like
<code class="highlighter-rouge">awk '/&gt;/ {print "&gt;A_meyeri_W712_ctg" i++} !/&gt;/ {print $0}' A_W712.fasta &gt; A_W712-renamed.fasta</code>.
The original genomes went into a directory called <code class="highlighter-rouge">original_genomes</code> and
the renamed files into <code class="highlighter-rouge">genomes</code>.</p>

<p><strong>Side note:</strong> We ran all of these analyses on an 8-core 2019 MacBook
Pro with 32GB RAM, except for the functional annotation with
Interproscan, which we ran on Harvard University’s Odyssey cluster to
make that step faster. This cluster uses the Slurm job scheduler; we
have included the header information for the sake of repeatability and
to provide context about the resources used and the means of
parallelization.</p>

<h2 id="pangenome-construction">Pangenome construction</h2>

<p>To relate the genomes based on gene content and investigate potentially
shared distinctive features, we made a pangenome. We used
<a href="https://github.com/merenlab/anvio/tree/master/anvio">anvi’o</a>, a
framework for analysis and visualization of ’omics data, for the
majority of this project.</p>

<p>First we added all our genomes’ contigs (contiguous chunks) into a
single contigs database. First we combined all the renamed fastas, like
<code class="highlighter-rouge">cat genomes/*renamed.fasta &gt; Actinomyces-23.fa</code>. Then we converted this
to an anvi’o contigs database with</p>

<p><code class="highlighter-rouge">anvi-gen-contigs-database -f Actinomyces-23.fa -o Actinomyces-23-CONTIGS.db</code></p>

<p>During this step, anvi’o invoked
<a href="https://github.com/hyattpd/Prodigal">Prodigal</a> to call genes on our
contigs.</p>

<h2 id="annotating-gene-calls">Annotating gene calls</h2>

<p>The predicted genes were then annotated using
<a href="https://www.ebi.ac.uk/interpro/about/interpro/">Interproscan</a> that is a
useful bundle of a variety of annotation sources like Pfam and Tigrfam.
To annotate the genomes rather quickly, we parallelized the process by
choping the job up into multiple batches which ran concurrently on our
university’s cluster. To get the gene sequences (amino acid), we ran</p>

<p><code class="highlighter-rouge">anvi-get-sequences-for-gene-calls -c Actinomyces-23-CONTIGS.db --wrap 0 --get-aa-sequences -o Actinomyces-23-gene-calls.faa</code></p>

<p>This was chopped up into chunks of 5000 genes apiece by
<code class="highlighter-rouge">split -l 10000 -d -a 4 Actinomyces-23-gene-calls.faa Actinomyces-23-gene-calls.faa-</code>
and then we set up the column names of the output table with
<code class="highlighter-rouge">echo -e "gene_callers_id\tsource\taccession\tfunction\te_value" &gt; interpro-results-formatted-Actinomyces-23.tsv</code></p>

<p>The split command generated 11 files starting of the form
<em>PREFIX.faa-BATCH</em> (e.g. <code class="highlighter-rouge">Actinomyces-23-gene-calls.faa-0004</code>), which
fed into the following Slurm job array script:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>
<span class="c">#SBATCH -N 1 #1 nodes of ram</span>
<span class="c">#SBATCH -n 12 # 12 cores from each</span>
<span class="c">#SBATCH --contiguous</span>
<span class="c">#SBATCH --mem=12G #per node</span>
<span class="c">#SBATCH -t 0-6:00:00</span>
<span class="c">#SBATCH -p shared</span>
<span class="c">#SBATCH --array=0-10%11</span>
<span class="c">#SBATCH --job-name="anvi-ipAct"</span>
<span class="c">#SBATCH -o odyssey_anviIPact.out</span>
<span class="c">#SBATCH -e odyssey_anviIPact.err</span>
<span class="c">#SBATCH --mail-type=END</span>

<span class="c"># set this up first</span>
<span class="c"># split -l 10000 -d -a 4 Actinomyces-23-gene-calls.faa Actinomyces-23-gene-calls.faa-</span>
<span class="c"># echo -e "gene_callers_id\tsource\taccession\tfunction\te_value" &gt; interpro-results-formatted-Actinomyces-23.tsv</span>


<span class="nv">prefix</span><span class="o">=</span><span class="s2">"Actinomyces-23"</span>
<span class="nv">taskNum</span><span class="o">=</span><span class="si">$(</span><span class="nb">printf</span> %04d <span class="nv">$SLURM_ARRAY_TASK_ID</span><span class="si">)</span>  <span class="c"># format slurm task id to be padded with 0s like split</span>

<span class="c"># identify which batch this array element matches</span>
<span class="nv">batch</span><span class="o">=</span><span class="nv">$prefix</span><span class="nt">-gene-calls</span>.faa-<span class="nv">$taskNum</span>

<span class="c"># load java for interproscan</span>
module load jdk/1.8.0_172-fasrc01

<span class="c"># run interproscan on this batch</span>
./../../interproscan-sh/interproscan-5.36-75.0/interproscan.sh <span class="nt">-i</span> <span class="nv">$batch</span> <span class="nt">-o</span> <span class="k">${</span><span class="nv">batch</span><span class="k">}</span>.tsv <span class="nt">-f</span> tsv <span class="nt">--appl</span> TIGRFAM,Pfam,SUPERFAMILY,ProDom,Gene3D

<span class="c"># format it manually and append to main table</span>
<span class="nb">cat</span> <span class="k">${</span><span class="nv">batch</span><span class="k">}</span>.tsv | <span class="nb">awk</span> <span class="nt">-F</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span> <span class="s1">'{print $1 FS $4 FS $5 FS $6 FS $9}'</span> <span class="o">&gt;&gt;</span> interpro-results-formatted-<span class="nv">$prefix</span>.tsv
</code></pre></div></div>

<p>Once this finished, we imported the results into the contigs database
with
<code class="highlighter-rouge">anvi-import-functions -c Actinomyces-23-CONTIGS.db -i interpro-results-formatted-Actinomyces-23.tsv</code></p>

<p>We also ran HMM profiles for a custom set of canonically single copy
core genes anvi’o knows about, with
<code class="highlighter-rouge">anvi-run-hmms -c Actinomyces-23-CONTIGS.db -T 8</code> (the <code class="highlighter-rouge">-T 8</code> used 8
threads to speed things up). From this information anvi’o will later be
able to estimate each genome’s completeness and redundancy, based on the
number of single copy core genes missing or duplicated, respectively.</p>

<p>Then we made a dummy profile collection since that is required for the
next step by anvi’o, like so:</p>

<p><code class="highlighter-rouge">anvi-profile --blank-profile -c Actinomyces-23-CONTIGS.db -o Actinomyces-23-PROFILE -S dummy</code></p>

<p>The main purpose of this is to hold a ‘collection’ that tells anvi’o
which contigs came from which genome. To do this anvi’o needs a file
with two columns where the left column is the contig name and right
column is the collection name (here, genome name) to which that contig
belongs.</p>

<p>We accomplished this with a quick bash oneliner to trim off the <code class="highlighter-rouge">_ctg#</code>
suffix from the contig deflines, like</p>

<p><code class="highlighter-rouge">awk -F"_ctg" '/&gt;/ {print $0 "\t" $1}' Actinomyces-23.fa | sed 's/&gt;//g' &gt; collection_mapper_Actinomyces-23.txt</code></p>

<p>Then we can move the profile database to an convenient place like
<code class="highlighter-rouge">mv Actinomyces-23-PROFILE/PROFILE.db Actinomyces-23-PROFILE.db</code>, then
import the file into the PROFILE db to associate the contigs into
genomes:</p>

<p><code class="highlighter-rouge">anvi-import-collection -p Actinomyces-23-PROFILE.db -c Actinomyces-23-CONTIGS.db -C Genomes collection_mapper_Actinomyces-23.txt --contigs-mode</code></p>

<p>With this information in hand we can then write a file that contains all
the info for how to identify each genome and where to find its data,
which we call <code class="highlighter-rouge">Actinomyces-23-internal-genomes-table.txt</code> whose contents
are copied here for convenience:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>name    bin_id  collection_id   profile_db_path contigs_db_path
A_odontolyticus_ATCC_17929  A_odontolyticus_ATCC_17929  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_odontolyticus_ATCC_17982  A_odontolyticus_ATCC_17982  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_meyeri_ATCC_35568 A_meyeri_ATCC_35568 Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_odontolyticus_F0309   A_odontolyticus_F0309   Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_180_F0310  A_sp_HMT_180_F0310  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_172_F0311  A_sp_HMT_172_F0311  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_849_F0330  A_sp_HMT_849_F0330  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_848_F0332  A_sp_HMT_848_F0332  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_171_F0337  A_sp_HMT_171_F0337  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_178_F0338  A_sp_HMT_178_F0338  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_175_F0384  A_sp_HMT_175_F0384  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_170_F0386  A_sp_HMT_170_F0386  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_448_F0400  A_sp_HMT_448_F0400  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_massiliensis_F0489    A_massiliensis_F0489    Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_johnsonii_F0510   A_johnsonii_F0510   Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_graevenitzii_F0530    A_graevenitzii_F0530    Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_877_F0543  A_sp_HMT_877_F0543  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_HMT_414_F0588  A_sp_HMT_414_F0588  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_ICM39  A_sp_ICM39  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_ICM47  A_sp_ICM47  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_sp_ICM58  A_sp_ICM58  Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_meyeri_W712   A_meyeri_W712   Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
A_odontolyticus_XH001   A_odontolyticus_XH001   Genomes Actinomyces-23-PROFILE.db   Actinomyces-23-CONTIGS.db
</code></pre></div></div>

<p>Now we can pull the relevant information for these genomes using the
following anvi’o command:</p>

<p><code class="highlighter-rouge">anvi-gen-genomes-storage -i Actinomyces-23-internal-genomes-table.txt -o Actinomyces-23-GENOMES.db</code></p>

<p>And then use the resultant GENOMES.db file to calculate the pangenome:
<code class="highlighter-rouge">anvi-pan-genome -g Actinomyces-23-GENOMES.db -o Actinomyces-23-PAN -n Actinomyces-23-mcl6 --mcl-inflation 6 --use-ncbi-blast -T 10</code></p>

<p>This step is where we operationally define gene homology. To do this,
anvi’o first uses blastp (<code class="highlighter-rouge">--use-ncbi-blast</code>) to compute the amino acid
similarities between each pair of genes (technically, predicted ORFs).
Then, the blastp similarities are used to generate a network linking
genes, where edges are blastp similarities and nodes are genes.
<a href="https://micans.org/mcl/">MCL</a>, a network clustering algorithm, is then
applied to this network to identify ‘clusters’ of genes on the graph,
that represent genes clustered together in amino-acid space. These
operationally defined gene homologs are so called <strong>gene clusters</strong>. The
tendency of MCL to split or lump clusters is based on the hyperparameter
<code class="highlighter-rouge">--mcl-inflation</code> which we set to 6, as that has produced good results
for genus-level pangenomes.</p>

<h2 id="adding-some-miscellaneous-data">Adding some miscellaneous data</h2>

<p>Once the pangenome was done, we imported information about the number of
contigs and the TM7x susceptibility of each genome. We found the number
of contigs for each genome by running
<code class="highlighter-rouge">awk '{print $2}' collection_mapper_Actinomyces-23.txt | uniq -c</code> and
then manually adding the corresponding TM7x information from the wet
experiments. This resulted in the following table:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>layer   Hosting NumCtgs
A_odontolyticus_ATCC_17929  Crash   26
A_odontolyticus_ATCC_17982  Crash   2
A_meyeri_ATCC_35568 Crash   15
A_odontolyticus_F0309   Crash   6
A_sp_HMT_180_F0310  Crash   6
A_sp_HMT_172_F0311  No_crash    215
A_sp_HMT_849_F0330  Resistant   76
A_sp_HMT_848_F0332  Resistant   1
A_sp_HMT_171_F0337  Resistant   280
A_sp_HMT_178_F0338  Crash   43
A_sp_HMT_175_F0384  Resistant   7
A_sp_HMT_170_F0386  Resistant   76
A_sp_HMT_448_F0400  Resistant   12
A_massiliensis_F0489    Resistant   233
A_johnsonii_F0510   Resistant   324
A_graevenitzii_F0530    Resistant   29
A_sp_HMT_877_F0543  Crash   1
A_sp_HMT_414_F0588  Resistant   1
A_sp_ICM39  Crash   146
A_sp_ICM47  No_crash    143
A_sp_ICM58  No_crash    1
A_meyeri_W712   Crash   1
A_odontolyticus_XH001   Crash   5
</code></pre></div></div>

<p>This was imported into the pangenome for visual display with</p>

<p><code class="highlighter-rouge">anvi-import-misc-data -p Actinomyces-23-PAN/Actinomyces-23-mcl6-PAN.db -t layers Actinomyces-23-external-table.txt</code></p>

<p>We also converted the susceptibility information to a numeric variable
(resistant=0, No_crash=0.5, Crash=1) and similarly imported it.</p>

<p>Altogether, this generated the pangenome displayed in Figure 5, when
colored and formatted, and the genomes are oriented by gene cluster
frequencies.</p>

<h1 id="enriched-functions">Enriched functions</h1>

<p>Based on the layout in the pangenome and the gene cluster that were
visibly core to each group, we wondered if we could identify functions
enriched in each group. To that end, we applied a built-in function in
anvi’o that ultimately compares genomes based on what predicted
functions they share or don’t share, rather than specific homologs (as
is displayed in the pangenome) (Shaiber et al., submitted). This method
works by passing anvi’o a list of which genomes belong in which
category, a functional annotation (here, Pfam), and then anvi’o counts
up, for each function, the number of genomes by category. From this, a
table is made that lists each function, its ‘enrichment score’
reflecting the ratio of in- vs. out-of-category genomes having that
predicted function, a p-value and FDR-adjusted q-value, and the
proportion of genomes in each category containing that function (so we
biologists can decide whether we believe the machine’s estimate of
significant enrichment). All this is done with the following command:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>anvi-get-enriched-functions-per-pan-group -p Actinomyces-23-PAN/Actinomyces-23-mcl6-PAN.db -g Actinomyces-23-GENOMES.db --category Hosting --annotation-source Pfam -o Actinomyces-23-ENRICHED-PFAM.txt
</code></pre></div></div>

<p>This generated Table S4 of the manuscript, from which the top genes in
relevant categories were taken to report in the in-text Table 1.</p>

<h2 id="investigating-single-gene-trees-from-core-gene-clusters">Investigating single-gene trees from core gene clusters</h2>

<p>While presence/absence of gene clusters or functions may distinguish
features of <em>Actinomyces</em> strains important determining TM7x response,
this method assumes that the phenotype is based on the complete presence
or absence of a gene/function. However, it is biologically plausible
(and evident in many host/phage systems) that a conserved gene may be
the target, in which case seemingly minor amino-acid sequence variants
can effect a strong change in the host/symbiont relationship. To address
this second scenario, we decided to look for amino acid sequence
variants in gene clusters core to all 23 genomes, or core to the 13
genomes that could host TM7x. Specifically, we looked for variants that
distinguished the various groups (e.g. non-crashers vs crashers vs
resistant).</p>

<h2 id="extracting-genes-and-generating-gene-trees">Extracting genes and generating gene trees</h2>

<p>So first we exported all gene clusters that were present in all 23 or
the 13 TM7x-hosting genomes, but only gene clusters represented by a
single gene sequence per genome. We accomplished that by first
summarizing the pangenome so that we record in the pangenome database
which gene clusters occur in which core sets (e.g. core to all genomes,
core to resistant, etc):
<code class="highlighter-rouge">anvi-summarize -g Actinomyces-23-GENOMES.db -p Actinomyces-23-PAN/Actinomyces-23-mcl6-PAN.db -C cores -o Actinomyces-23-SUMMARY</code></p>

<p>followed by
<code class="highlighter-rouge">gunzip Actinomyces-23-SUMMARY/Actinomyces-23-mcl6_gene_clusters_summary.txt.gz</code></p>

<p>And the resultant <code class="highlighter-rouge">*summary.txt</code> file contains information for all gene
clusters and genomes in the pangenome, which is Table S3 in the
manuscript.</p>

<p>We then pulled out the gene clusters of interest:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>gc <span class="k">in</span> <span class="si">$(</span><span class="nb">grep</span> <span class="s1">'Susceptible_core'</span> Actinomyces-23-SUMMARY/Actinomyces-23-mcl6_gene_clusters_summary.txt | <span class="nb">awk</span> <span class="nt">-F</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span> <span class="s1">'{print $2}'</span> | <span class="nb">uniq</span><span class="si">)</span><span class="p">;</span> <span class="k">do

</span>anvi-get-sequences-for-gene-clusters <span class="nt">-p</span> Actinomyces-23-PAN/Actinomyces-23-mcl6-PAN.db <span class="nt">-g</span> Actinomyces-23-GENOMES.db <span class="se">\</span>
<span class="nt">--max-num-genes-from-each-genome</span> 1 <span class="nt">--gene-cluster-id</span> <span class="nv">$gc</span> <span class="nt">-o</span> susceptible_core_gcs/<span class="nv">$gc</span>.faa

<span class="k">done</span>
</code></pre></div></div>

<p>And</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>gc <span class="k">in</span> <span class="si">$(</span><span class="nb">grep</span> <span class="s1">'All_core'</span> Actinomyces-23-SUMMARY/Actinomyces-23-mcl6_gene_clusters_summary.txt | <span class="nb">awk</span> <span class="nt">-F</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span> <span class="s1">'{print $2}'</span> | <span class="nb">uniq</span><span class="si">)</span><span class="p">;</span> <span class="k">do

</span>anvi-get-sequences-for-gene-clusters <span class="nt">-p</span> Actinomyces-23-PAN/Actinomyces-23-mcl6-PAN.db <span class="nt">-g</span> Actinomyces-23-GENOMES.db <span class="se">\</span>
<span class="nt">--max-num-genes-from-each-genome</span> 1 <span class="nt">--gene-cluster-id</span> <span class="nv">$gc</span> <span class="nt">-o</span> all_core_gcs/<span class="nv">$gc</span>.faa

<span class="k">done</span>
</code></pre></div></div>

<p>This identified and saved alignments for 419 gene clusters for the 13
susceptible strains and 291 gene clusters for all 23 genomes.</p>

<p>The deflines contained more information than we wanted in the tip labels
so we simplified with <code class="highlighter-rouge">sed -i "" 's/&gt;.*genome_name:/&gt;/; s/|.*$//' *faa</code></p>

<p>Then we generated phylogenies with
<a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009490">FastTree</a>
for each gene with <code class="highlighter-rouge">for gc in $(ls); do FastTree $gc &gt; $gc.nwk; done</code>
which produced a newick tree for each gene cluster.</p>

<h2 id="identifying-gene-clusters-producing-interesting-topologies">Identifying gene clusters producing interesting topologies</h2>

<p>Now that we had all these gene trees, we screened through them quickly
with a custom python script:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">Tree</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">combinations</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>

<span class="c1"># this is the dendrogram produced by gene cluster frequency in pangenome in Fig 5, also cartoonized in Fig 6a
</span><span class="n">refAll</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="s">"(((A_sp_HMT_848_F0332,A_graevenitzii_F0530),(((A_sp_HMT_448_F0400,A_massiliensis_F0489),</span><span class="se">\
</span><span class="s">A_sp_HMT_414_F0588),(((A_sp_HMT_175_F0384,A_sp_HMT_171_F0337),A_sp_HMT_170_F0386),</span><span class="se">\
</span><span class="s">(A_sp_HMT_849_F0330,A_johnsonii_F0510)))),(((((A_sp_ICM58,A_sp_HMT_172_F0311),A_sp_ICM47),</span><span class="se">\
</span><span class="s">((((A_odontolyticus_F0309,A_odontolyticus_ATCC_17982),A_odontolyticus_ATCC_17929),A_sp_HMT_180_F0310),</span><span class="se">\
</span><span class="s">(A_sp_ICM39,A_odontolyticus_XH001))),(A_meyeri_W712,A_meyeri_ATCC_35568)),(A_sp_HMT_877_F0543,A_sp_HMT_178_F0338)));"</span><span class="p">)</span>

<span class="c1"># Define tips belonging to the various clades
</span><span class="n">noCrashClade</span> <span class="o">=</span> <span class="p">[</span><span class="s">'A_sp_ICM58'</span><span class="p">,</span> <span class="s">'A_sp_HMT_172_F0311'</span><span class="p">,</span> <span class="s">'A_sp_ICM47'</span><span class="p">]</span>
<span class="n">crashersInternal</span> <span class="o">=</span> <span class="p">[</span><span class="s">'A_odontolyticus_F0309'</span><span class="p">,</span><span class="s">'A_odontolyticus_ATCC_17982'</span><span class="p">,</span><span class="s">'A_odontolyticus_ATCC_17929'</span><span class="p">,</span>
                    <span class="s">'A_sp_HMT_180_F0310'</span><span class="p">,</span><span class="s">'A_sp_ICM39'</span><span class="p">,</span><span class="s">'A_odontolyticus_XH001'</span><span class="p">]</span>
<span class="n">crashersPolyphyletic</span> <span class="o">=</span> <span class="p">[</span><span class="s">'A_meyeri_W712'</span><span class="p">,</span><span class="s">'A_meyeri_ATCC_35568'</span><span class="p">,</span><span class="s">'A_sp_HMT_877_F0543'</span><span class="p">,</span><span class="s">'A_sp_HMT_178_F0338'</span><span class="p">]</span>
<span class="n">resistant</span> <span class="o">=</span> <span class="p">[</span><span class="s">'A_sp_HMT_848_F0332'</span><span class="p">,</span><span class="s">'A_graevenitzii_F0530'</span><span class="p">,</span><span class="s">'A_sp_HMT_448_F0400'</span><span class="p">,</span><span class="s">'A_massiliensis_F0489'</span><span class="p">,</span>
             <span class="s">'A_sp_HMT_414_F0588'</span><span class="p">,</span><span class="s">'A_sp_HMT_175_F0384'</span><span class="p">,</span><span class="s">'A_sp_HMT_171_F0337'</span><span class="p">,</span><span class="s">'A_sp_HMT_170_F0386'</span><span class="p">,</span><span class="s">'A_sp_HMT_849_F0330'</span><span class="p">,</span>
             <span class="s">'A_johnsonii_F0510'</span><span class="p">]</span>

<span class="n">comparisons</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># to save a record matching gene clusters with their type of topology
</span><span class="n">usefulTrees</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># to save the associated trees
</span>
<span class="c1"># For the 23 core set
</span><span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s">"/Users/dutter/g4/bat_tm7/core_gene_clusters/*nwk"</span><span class="p">):</span>

    <span class="n">t</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="c1"># read tree
</span>    <span class="c1"># Course check for whether non-crashers and resistant are each monophyletic
</span>    <span class="k">if</span> <span class="n">t</span><span class="o">.</span><span class="n">check_monophyly</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">noCrashClade</span><span class="p">,</span> <span class="n">target_attr</span><span class="o">=</span><span class="s">'name'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">and</span> \
            <span class="n">t</span><span class="o">.</span><span class="n">check_monophyly</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">resistant</span><span class="p">,</span> <span class="n">target_attr</span><span class="o">=</span><span class="s">'name'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]:</span>

        <span class="c1"># extract the gene cluster ID from the filename
</span>        <span class="n">gene_cluster_id</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">r"^.*/(.*).faa.nwk"</span><span class="p">,</span> <span class="s">'</span><span class="se">\\</span><span class="s">1'</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>  

        <span class="c1"># Now check ideal case - non-crashers are sister to resistant, together sister to the crashers - Fig 6c
</span>        <span class="k">if</span> <span class="n">t</span><span class="o">.</span><span class="n">check_monophyly</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">noCrashClade</span> <span class="o">+</span> <span class="n">resistant</span><span class="p">,</span> <span class="n">target_attr</span><span class="o">=</span><span class="s">'name'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">and</span> \
                <span class="n">t</span><span class="o">.</span><span class="n">check_monophyly</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">crashersInternal</span><span class="p">,</span> <span class="n">target_attr</span><span class="o">=</span><span class="s">'name'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="n">comparisons</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span> <span class="o">=</span> <span class="s">'(((noCrash) + (resistant)) + (most_crashers))'</span>
            <span class="n">usefulTrees</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span>

        <span class="c1"># but also interested if all three groups are monophyletic but we don't specify the relationship between them - Fig 6b
</span>        <span class="k">elif</span> <span class="nb">any</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">check_monophyly</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">crashersInternal</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">combo</span><span class="p">),</span> <span class="n">target_attr</span><span class="o">=</span><span class="s">'name'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                  <span class="k">for</span> <span class="n">length</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span> <span class="k">for</span> <span class="n">combo</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="n">crashersPolyphyletic</span><span class="p">,</span> <span class="n">length</span><span class="p">)]):</span>
            <span class="n">comparisons</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span><span class="o">=</span><span class="s">'no_crash, resistant, most non_crashers each monophyletic but variably related'</span>
            <span class="n">usefulTrees</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span>


<span class="c1"># do likewise for the gene clusters core to 13 susceptible genomes - Fig 6d
</span><span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s">"/Users/dutter/g4/bat_tm7/susceptible_core_gcs/*nwk"</span><span class="p">):</span>

    <span class="n">t</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>  <span class="c1"># read tree in
</span>            
    <span class="c1"># no resistant genomes so only check if non-crashers and the odontolyticus group of 6 genomes are each monophyletic regardless of relationship
</span>    <span class="k">if</span> <span class="n">t</span><span class="o">.</span><span class="n">check_monophyly</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">noCrashClade</span><span class="p">,</span> <span class="n">target_attr</span><span class="o">=</span><span class="s">'name'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">and</span> \
            <span class="nb">any</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">check_monophyly</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">crashersInternal</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">combo</span><span class="p">),</span> <span class="n">target_attr</span><span class="o">=</span><span class="s">'name'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                 <span class="k">for</span> <span class="n">length</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span> <span class="k">for</span> <span class="n">combo</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="n">crashersPolyphyletic</span><span class="p">,</span> <span class="n">length</span><span class="p">)]):</span>

        <span class="c1"># extract the gene cluster ID from the filename
</span>        <span class="n">gene_cluster_id</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">r"^.*/(.*).faa.nwk"</span><span class="p">,</span> <span class="s">'</span><span class="se">\\</span><span class="s">1'</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>  <span class="c1"># extract 
</span>
        <span class="n">comparisons</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span> <span class="o">=</span> <span class="s">'(noCrash) + (most_crashers)'</span>
        <span class="n">usefulTrees</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span>


<span class="c1"># read in the gene cluster summary information to get functions
</span><span class="n">pan</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"~/g4/bat_tm7/Actinomyces-23-SUMMARY/Actinomyces-23-mcl6_gene_clusters_summary.txt"</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s">'str'</span><span class="p">)</span>

<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s">'display.max_columns'</span><span class="p">,</span> <span class="mi">7</span><span class="p">)</span>

<span class="c1"># subset summary table to only gene clusters identified here and columns of interest
</span><span class="n">subPan</span> <span class="o">=</span> <span class="n">pan</span><span class="p">[</span><span class="n">pan</span><span class="p">[</span><span class="s">'gene_cluster_id'</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">comparisons</span><span class="o">.</span><span class="n">keys</span><span class="p">()))]</span>
<span class="n">subPan</span> <span class="o">=</span> <span class="n">subPan</span><span class="p">[[</span><span class="s">'gene_cluster_id'</span><span class="p">,</span><span class="s">'Pfam'</span><span class="p">,</span><span class="s">'TIGRFAM'</span><span class="p">,</span><span class="s">'Preferred_Name'</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

<span class="c1"># add a column called treetype that describes the relationship for that gene
</span><span class="n">subPan</span><span class="p">[</span><span class="s">'treetype'</span><span class="p">]</span> <span class="o">=</span> <span class="n">subPan</span><span class="p">[</span><span class="s">'gene_cluster_id'</span><span class="p">]</span><span class="o">.</span><span class="nb">map</span><span class="p">(</span><span class="n">comparisons</span><span class="p">)</span>

<span class="k">print</span><span class="p">(</span><span class="n">subPan</span><span class="p">)</span>

<span class="c1">#subPan.to_csv('~/g4/bat_tm7/Actinomyces-23-gene-tree-results.tsv', sep="\t")  # turned off here for example
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
          gene_cluster_id                                               Pfam  \
    3168      GC_00000129                 MerR HTH family regulatory protein   
    3398      GC_00000139  Ribosomal L18 of archaea, bacteria, mitoch. an...   
    4111      GC_00000170                                         HIT domain   
    5721      GC_00000240                                    ABC transporter   
    7975      GC_00000338                              SecA DEAD-like domain   
    8527      GC_00000362                               Peptidase family M41   
    18005     GC_00000930               Haloacid dehalogenase-like hydrolase   
    18007     GC_00000930               Haloacid dehalogenase-like hydrolase   
    18811     GC_00000992                                          G5 domain   
    19344     GC_00001033                              Pterin binding enzyme   
    19669     GC_00001058                           Prephenate dehydrogenase   
    19678     GC_00001058                                                NaN   
    20202     GC_00001099                                  Cytidylate kinase   
    20423     GC_00001116                         Aldo/keto reductase family   
    20436     GC_00001117                         4-alpha-glucanotransferase   
    20644     GC_00001133                L,D-transpeptidase catalytic domain   
    21996     GC_00001237                                       NUDIX domain   

                                                 TIGRFAM Preferred_Name  \
    3168                                             NaN          merR2   
    3398                L18_bact: ribosomal protein uL18           rplR   
    4111                                             NaN            hit   
    5721                                             NaN           ugpC   
    7975      secA: preprotein translocase, SecA subunit           secA   
    8527   FtsH_fam: ATP-dependent metallopeptidase HflB           ftsH   
    18005                                            NaN           yutF   
    18007          HAD-SF-IIA: HAD hydrolase, family IIA           yutF   
    18811                                            NaN            NaN   
    19344                 DHPS: dihydropteroate synthase           folP   
    19669                                            NaN           tyrA   
    19678                                            NaN           tyrA   
    20202                         cmk: cytidylate kinase            cmk   
    20423                                            NaN            NaN   
    20436               malQ: 4-alpha-glucanotransferase           malQ   
    20644                                            NaN         enhA_2   
    21996                                            NaN           mutT   

                                                    treetype  
    3168   no_crash, resistant, most non_crashers each mo...  
    3398       (((noCrash) + (resistant)) + (most_crashers))  
    4111   no_crash, resistant, most non_crashers each mo...  
    5721       (((noCrash) + (resistant)) + (most_crashers))  
    7975   no_crash, resistant, most non_crashers each mo...  
    8527   no_crash, resistant, most non_crashers each mo...  
    18005                        (noCrash) + (most_crashers)  
    18007                        (noCrash) + (most_crashers)  
    18811                        (noCrash) + (most_crashers)  
    19344                        (noCrash) + (most_crashers)  
    19669                        (noCrash) + (most_crashers)  
    19678                        (noCrash) + (most_crashers)  
    20202                        (noCrash) + (most_crashers)  
    20423                        (noCrash) + (most_crashers)  
    20436                        (noCrash) + (most_crashers)  
    20644                        (noCrash) + (most_crashers)  
    21996                        (noCrash) + (most_crashers)  

And this is the data for Figure 6b-d. To put these few genes displaying
this topology in context, the diversity of tree topologies can be
visualized with a density tree overlaying the hundreds of gene trees.
First the genes are concatenated into a multiphylo file, with the tree
name in the first column and the corresponding newick tree in the second
column:

``` bash
for file in $(ls core_gene_clusters/GC_0000*nwk); do 
  # get the gene cluster ID from filename
  gc=$(echo $file | sed 's,^.*/,,; s,\..*$,,')
  # print ID, space, newick tree for this tree
  sed "s/^/$gc /" $file 
done &gt; core_gene_clusters.nwk
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>file <span class="k">in</span> <span class="si">$(</span><span class="nb">ls </span>susceptible_core_gcs/GC_0000<span class="k">*</span>nwk<span class="si">)</span><span class="p">;</span> <span class="k">do</span> 
  <span class="c"># get the gene cluster ID from filename</span>
  <span class="nv">gc</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$file</span> | <span class="nb">sed</span> <span class="s1">'s,^.*/,,; s,\..*$,,'</span><span class="si">)</span>
  <span class="c"># print ID, space, newick tree for this tree</span>
  <span class="nb">sed</span> <span class="s2">"s/^/</span><span class="nv">$gc</span><span class="s2"> /"</span> <span class="nv">$file</span> 
<span class="k">done</span> <span class="o">&gt;</span> susceptible_core_gcs.nwk
</code></pre></div></div>

<p>Then we can overlay all the newick trees for the 291 genes core to all
23 genomes:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">phangorn</span><span class="p">)</span><span class="w">

</span><span class="n">allCore</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.tree</span><span class="p">(</span><span class="s2">"core_gene_clusters.nwk"</span><span class="p">)</span><span class="w">
</span><span class="n">densiTree</span><span class="p">(</span><span class="n">allCore</span><span class="p">,</span><span class="w"> </span><span class="n">consensus</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">consensus</span><span class="p">(</span><span class="n">allCore</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'cladogram'</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="n">use.edge.length</span><span class="o">=</span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">scaleX</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">scale.bar</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="Act23_phylo_files/figure-markdown_github/unnamed-chunk-10-1.png" alt="" /></p>

<p>And the 419 gene clusters core to the 13 susceptible strains:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">susceptibleCore</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.tree</span><span class="p">(</span><span class="s2">"susceptible_core_gcs.nwk"</span><span class="p">)</span><span class="w">
</span><span class="n">densiTree</span><span class="p">(</span><span class="n">susceptibleCore</span><span class="p">,</span><span class="w"> </span><span class="n">consensus</span><span class="o">=</span><span class="n">consensus</span><span class="p">(</span><span class="n">susceptibleCore</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'cladogram'</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="n">use.edge.length</span><span class="o">=</span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">scaleX</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">scale.bar</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="Act23_phylo_files/figure-markdown_github/unnamed-chunk-12-1.png" alt="" /></p>

<h2 id="phylogenetic-analysis-1-amino-acid-identity-aai">Phylogenetic analysis 1: Amino Acid Identity (AAI)</h2>

<p>To get an idea of how different our genomes are, phylogenetically, since
many represent unnamed species, we compared average amino acid identity
(AAI). This metric works by calling genes with Prodigal, comparing
translated AA sequences with
<a href="https://github.com/bbuchfink/diamond">Diamond</a> for all gene pairs for
each genome. For context, <a href="https://academic.oup.com/nar/article/42/8/e73/1076763">Luo et
al. 2014</a> have a
nice Figure 2 showing % AAI for genomes across various taxomonic levels.
Species and genus have fairly broad AAI distributions, but all species
are above 80% AAI with the majority being above 90%, while congeners are
in the 50-85% range.</p>

<p>We used <a href="https://github.com/dparks1134/CompareM">CompareM</a> to compute
AAI using their default workflow <code class="highlighter-rouge">aai_wf</code>, which we invoked like so:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>comparem aai_wf genomes genomes_aai --cpus 8 --file_ext .fasta
</code></pre></div></div>

<p>The key output was a file <code class="highlighter-rouge">genomes_aai/aai/aai_summary.tsv</code> that
contains the AAI information for each genome. We then modified <a href="https://hackmd.io/L2llRUU_SrWfI4OYN-uozQ?view#Convert-output-into-a-sourmash-compare-style-numpy-matrix">Titus
Brown’s python
script</a>
to parse the output into a nice table, our modified file copied here:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#! /usr/bin/env python
</span><span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span><span class="p">,</span> <span class="n">division</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">argparse</span>

<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">'aai_summary_tsv'</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">'-o'</span><span class="p">,</span> <span class="s">'--output'</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="k">assert</span> <span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s">'please specify --output'</span>

    <span class="n">indices</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">aai_summary_tsv</span><span class="p">,</span> <span class="s">'rt'</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="n">fp</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        
    <span class="n">lines</span> <span class="o">=</span> <span class="p">[</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lines</span> <span class="p">]</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="n">lines</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>

    <span class="c1"># assign unique indices to the thing
</span>    <span class="n">filenum</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
        <span class="n">g1</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">g2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">g1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">:</span>
            <span class="n">indices</span><span class="p">[</span><span class="n">g1</span><span class="p">]</span> <span class="o">=</span> <span class="n">filenum</span>
            <span class="n">filenum</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">g2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">:</span>
            <span class="n">indices</span><span class="p">[</span><span class="n">g2</span><span class="p">]</span> <span class="o">=</span> <span class="n">filenum</span>
            <span class="n">filenum</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">print</span><span class="p">(</span><span class="s">'...got {} genomes.'</span><span class="o">.</span><span class="nb">format</span><span class="p">(</span><span class="n">filenum</span><span class="p">))</span>

    <span class="n">D</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">filenum</span><span class="p">,</span> <span class="n">filenum</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
        <span class="n">g1</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">n1</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="n">g1</span><span class="p">]</span>
        <span class="n">g2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">n2</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="n">g2</span><span class="p">]</span>
        <span class="n">sim</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>

        <span class="n">D</span><span class="p">[</span><span class="n">n1</span><span class="p">,</span><span class="n">n2</span><span class="p">]</span> <span class="o">=</span> <span class="n">D</span><span class="p">[</span><span class="n">n2</span><span class="p">,</span><span class="n">n1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">sim</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">filenum</span><span class="p">):</span>
        <span class="n">D</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mf">100.0</span>

    <span class="n">x</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">indices</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">x</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">v</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
    <span class="n">x</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

    <span class="n">labeloutname</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">output</span> <span class="o">+</span> <span class="s">'.labels.txt'</span>
    <span class="k">print</span><span class="p">(</span><span class="s">'saving labels to: {}'</span><span class="o">.</span><span class="nb">format</span><span class="p">(</span><span class="n">labeloutname</span><span class="p">))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">labeloutname</span><span class="p">,</span> <span class="s">'w'</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
        <span class="n">fp</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">"</span><span class="se">\n</span><span class="s">"</span><span class="o">.</span><span class="n">join</span><span class="p">([</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">x</span> <span class="p">]))</span>

    <span class="k">print</span><span class="p">(</span><span class="s">'saving distance matrix to: {}'</span><span class="o">.</span><span class="nb">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s">'w'</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
        <span class="n">numpy</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">fp</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">z</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">z</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))</span>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">'__main__'</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</code></pre></div></div>

<p>We called this file <code class="highlighter-rouge">format_aai.py</code> and used it like
<code class="highlighter-rouge">./format_aai.py genomes_aai/aai/aai_summary.tsv -o genomes_aai/aai/aai_summary_fmt.tsv</code></p>

<p>The output was then manually formatted the last bit, producing the final
table we called <code class="highlighter-rouge">Actinomyces-23-aai.tsv</code> that we incorporated into
anvi’o, also copied here:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>layers  A_ICM47 A_F0309 A_ATCC17929 A_ATCC35568 A_F0588 A_ATCC17982 A_F0489 A_F0400 A_XH001 A_F0543 A_F0337 A_F0310 A_F0332 A_F0330 A_ICM58 A_F0338 A_W712  A_ICM39 A_F0386 A_F0384 A_F0530 A_F0510 A_F0311
A_ICM47 100 85.95   85.37   82.5    58.06   85.91   58.27   57.87   86.21   73.37   59.14   83.02   57.04   58.36   88.26   73.69   82.47   86.45   58.51   58.41   60.09   58.26   88.39
A_F0309 85.95   100 89.96   83.22   57.92   97.04   58.63   57.89   94.78   73.82   59.11   84.63   57.39   58.4    87.27   74.16   82.99   91.83   58.36   58.78   60.48   58.46   87.14
A_ATCC17929 85.37   89.96   100 83.07   58.01   89.98   58.43   57.94   89.42   73.71   59.29   84.53   57.24   58.57   86.58   74.41   83.01   90.63   58.48   58.68   59.92   58.7    86.37
A_ATCC35568 82.5    83.22   83.07   100 57.81   83.25   57.85   57.43   83.21   74.12   58.78   84.74   56.44   58.35   83.16   74.73   98.98   83.53   57.86   57.96   58.98   58.26   83.07
A_F0588 58.06   57.92   58.01   57.81   100 58.06   73.14   73.67   58.27   58.99   71.12   58.09   57.31   70.91   58.42   59.64   57.78   58.42   71.14   70.7    63.88   70.91   58.45
A_ATCC17982 85.91   97.04   89.98   83.25   58.06   100 58.45   58.11   94.88   73.85   59.25   84.43   57.31   58.48   87.42   74.32   83.28   91.9    58.23   58.78   60.36   58.4    87.35
A_F0489 58.27   58.63   58.43   57.85   73.14   58.45   100 77.68   58.46   58.91   71.95   58.57   57.3    71.43   58.73   59.59   57.98   58.72   71.51   71.98   64.36   71.48   58.61
A_F0400 57.87   57.89   57.94   57.43   73.67   58.11   77.68   100 58.21   58.47   71.26   57.86   56.9    71.44   58.41   59.07   57.42   58.03   71.45   71.16   63.89   71.42   58.04
A_XH001 86.21   94.78   89.42   83.21   58.27   94.88   58.46   58.21   100 73.72   59.41   84.41   57.25   58.72   87.77   74.43   83.24   93.04   58.54   58.76   60.11   58.65   87.7
A_F0543 73.37   73.82   73.71   74.12   58.99   73.85   58.91   58.47   73.72   100 59.84   74.25   57.15   59.31   74.33   97.71   74.1    74.03   58.99   59.25   59.55   59.07   74.05
A_F0337 59.14   59.11   59.29   58.78   71.12   59.25   71.95   71.26   59.41   59.84   100 59.26   58.04   87.01   59.63   60.29   58.85   59.26   87.89   91.43   65.63   87.33   59.47
A_F0310 83.02   84.63   84.53   84.74   58.09   84.43   58.57   57.86   84.41   74.25   59.26   100 56.94   58.56   83.98   74.67   84.48   84.44   58.22   58.7    59.76   58.55   83.5
A_F0332 57.04   57.39   57.24   56.44   57.31   57.31   57.3    56.9    57.25   57.15   58.04   56.94   100 57.89   57.46   57.79   56.61   57.23   57.41   57.44   57.27   57.8    57.48
A_F0330 58.36   58.4    58.57   58.35   70.91   58.48   71.43   71.44   58.72   59.31   87.01   58.56   57.89   100 58.75   59.95   58.32   58.68   88.49   86.75   64.95   97.22   58.88
A_ICM58 88.26   87.27   86.58   83.16   58.42   87.42   58.73   58.41   87.77   74.33   59.63   83.98   57.46   58.75   100 74.52   83.12   87.8    58.88   59.12   60.71   58.92   96.47
A_F0338 73.69   74.16   74.41   74.73   59.64   74.32   59.59   59.07   74.43   97.71   60.29   74.67   57.79   59.95   74.52   100 74.7    74.31   59.33   59.79   60.09   59.67   74.4
A_W712  82.47   82.99   83.01   98.98   57.78   83.28   57.98   57.42   83.24   74.1    58.85   84.48   56.61   58.32   83.12   74.7    100 83.55   58.01   58.07   58.82   58.25   82.91
A_ICM39 86.45   91.83   90.63   83.53   58.42   91.9    58.72   58.03   93.04   74.03   59.26   84.44   57.23   58.68   87.8    74.31   83.55   100 58.45   58.81   60.1    58.7    87.6
A_F0386 58.51   58.36   58.48   57.86   71.14   58.23   71.51   71.45   58.54   58.99   87.89   58.22   57.41   88.49   58.88   59.33   58.01   58.45   100 87.33   64.49   88.47   58.68
A_F0384 58.41   58.78   58.68   57.96   70.7    58.78   71.98   71.16   58.76   59.25   91.43   58.7    57.44   86.75   59.12   59.79   58.07   58.81   87.33   100 64.93   87.01   59.02
A_F0530 60.09   60.48   59.92   58.98   63.88   60.36   64.36   63.89   60.11   59.55   65.63   59.76   57.27   64.95   60.71   60.09   58.82   60.1    64.49   64.93   100 64.86   60.56
A_F0510 58.26   58.46   58.7    58.26   70.91   58.4    71.48   71.42   58.65   59.07   87.33   58.55   57.8    97.22   58.92   59.67   58.25   58.7    88.47   87.01   64.86   100 58.88
A_F0311 88.39   87.14   86.37   83.07   58.45   87.35   58.61   58.04   87.7    74.05   59.47   83.5    57.48   58.88   96.47   74.4    82.91   87.6    58.68   59.02   60.56   58.88   100
</code></pre></div></div>

<p>Genome names had gotten formatted differently between the analyses, so
we made a quick matching file called <code class="highlighter-rouge">matcher.txt</code> to synchronize the
formats:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>full_name   abbrev
A_odontolyticus_ATCC_17929  A_ATCC17929
A_odontolyticus_ATCC_17982  A_ATCC17982
A_meyeri_ATCC_35568 A_ATCC35568
A_odontolyticus_F0309   A_F0309
A_sp_HMT_180_F0310  A_F0310
A_sp_HMT_172_F0311  A_F0311
A_sp_HMT_849_F0330  A_F0330
A_sp_HMT_848_F0332  A_F0332
A_sp_HMT_171_F0337  A_F0337
A_sp_HMT_178_F0338  A_F0338
A_sp_HMT_175_F0384  A_F0384
A_sp_HMT_170_F0386  A_F0386
A_sp_HMT_448_F0400  A_F0400
A_massiliensis_F0489    A_F0489
A_johnsonii_F0510   A_F0510
A_graevenitzii_F0530    A_F0530
A_sp_HMT_877_F0543  A_F0543
A_sp_HMT_414_F0588  A_F0588
A_sp_ICM39  A_ICM39
A_sp_ICM47  A_ICM47
A_sp_ICM58  A_ICM58
A_meyeri_W712   A_W712
A_odontolyticus_XH001   A_XH001
</code></pre></div></div>

<p>We then plotted the AAI information like so:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">reshape2</span><span class="p">)</span><span class="w">

</span><span class="n">aai</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.csv</span><span class="p">(</span><span class="s2">"Actinomyces-23-aai.tsv"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">  </span><span class="c1"># matrix of aai values</span><span class="w">
</span><span class="n">matcher</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.csv</span><span class="p">(</span><span class="s2">"matcher.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">  </span><span class="c1"># to correct as some of the genome names don't match</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">matcher</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">matcher</span><span class="o">$</span><span class="n">abbrev</span><span class="w">  

</span><span class="n">aai</span><span class="o">$</span><span class="n">layers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">matcher</span><span class="p">[</span><span class="n">aai</span><span class="o">$</span><span class="n">layers</span><span class="p">,</span><span class="w"> </span><span class="s2">"full_name"</span><span class="p">]</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">aai</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">aai</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">matcher</span><span class="p">[</span><span class="n">colnames</span><span class="p">(</span><span class="n">aai</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">aai</span><span class="p">)],</span><span class="w"> </span><span class="s2">"full_name"</span><span class="p">])</span><span class="w">

</span><span class="c1"># this is the order from the pangenome dendrogram</span><span class="w">
</span><span class="n">panorder</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"A_sp_HMT_178_F0338"</span><span class="p">,</span><span class="s2">"A_sp_HMT_877_F0543"</span><span class="p">,</span><span class="s2">"A_meyeri_ATCC_35568"</span><span class="p">,</span><span class="s2">"A_meyeri_W712"</span><span class="p">,</span><span class="s2">"A_odontolyticus_XH001"</span><span class="p">,</span><span class="s2">"A_sp_ICM39"</span><span class="p">,</span><span class="s2">"A_sp_HMT_180_F0310"</span><span class="p">,</span><span class="s2">"A_odontolyticus_ATCC_17929"</span><span class="p">,</span><span class="s2">"A_odontolyticus_ATCC_17982"</span><span class="p">,</span><span class="s2">"A_odontolyticus_F0309"</span><span class="p">,</span><span class="s2">"A_sp_ICM47"</span><span class="p">,</span><span class="s2">"A_sp_HMT_172_F0311"</span><span class="p">,</span><span class="s2">"A_sp_ICM58"</span><span class="p">,</span><span class="s2">"A_johnsonii_F0510"</span><span class="p">,</span><span class="s2">"A_sp_HMT_849_F0330"</span><span class="p">,</span><span class="s2">"A_sp_HMT_170_F0386"</span><span class="p">,</span><span class="s2">"A_sp_HMT_171_F0337"</span><span class="p">,</span><span class="s2">"A_sp_HMT_175_F0384"</span><span class="p">,</span><span class="s2">"A_sp_HMT_414_F0588"</span><span class="p">,</span><span class="s2">"A_massiliensis_F0489"</span><span class="p">,</span><span class="s2">"A_sp_HMT_448_F0400"</span><span class="p">,</span><span class="s2">"A_graevenitzii_F0530"</span><span class="p">,</span><span class="s2">"A_sp_HMT_848_F0332"</span><span class="p">)</span><span class="w">

</span><span class="n">aaiLong</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">melt</span><span class="p">(</span><span class="n">aai</span><span class="p">)</span><span class="w">
</span><span class="c1"># order rows and columns to match the pangenome</span><span class="w">
</span><span class="n">aaiLong</span><span class="o">$</span><span class="n">layers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">aaiLong</span><span class="o">$</span><span class="n">layers</span><span class="p">),</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">rev</span><span class="p">(</span><span class="n">panorder</span><span class="p">)))</span><span class="w">
</span><span class="n">aaiLong</span><span class="o">$</span><span class="n">variable</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">aaiLong</span><span class="o">$</span><span class="n">variable</span><span class="p">),</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">rev</span><span class="p">(</span><span class="n">panorder</span><span class="p">)))</span><span class="w">

</span><span class="c1"># plot</span><span class="w">
</span><span class="n">ggplot</span><span class="p">(</span><span class="n">aaiLong</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">layers</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">variable</span><span class="p">,</span><span class="n">fill</span><span class="o">=</span><span class="n">value</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_tile</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="m">1.1</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">1.1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_text</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">'black'</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="nf">round</span><span class="p">(</span><span class="n">value</span><span class="p">,</span><span class="m">1</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">scale_y_discrete</span><span class="p">(</span><span class="n">expand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">scale_x_discrete</span><span class="p">(</span><span class="n">expand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">'Genome'</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="s1">'Genome'</span><span class="p">,</span><span class="n">fill</span><span class="o">=</span><span class="s1">'% AAI'</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">coord_equal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_gradientn</span><span class="p">(</span><span class="n">colours</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'steelblue3'</span><span class="p">,</span><span class="s1">'lightgoldenrodyellow'</span><span class="p">,</span><span class="w"> </span><span class="s1">'red'</span><span class="p">),</span><span class="w"> </span><span class="n">limits</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">60</span><span class="p">,</span><span class="m">100</span><span class="p">),</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'#cffacf'</span><span class="p">)</span><span class="o">+</span><span class="w"> </span><span class="n">theme_bw</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'black'</span><span class="p">),</span><span class="w"> </span><span class="n">axis.text.x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p><img src="Act23_phylo_files/figure-markdown_github/unnamed-chunk-17-1.png" alt="" /></p>

<h2 id="phylogenetic-analysis-2-phylophlan2">Phylogenetic analysis 2: PhyloPhlAn2</h2>

<p>For a more phylogenetic comparison of the genomes, we applied
<a href="https://bitbucket.org/nsegata/phylophlan/wiki/phylophlan2">PhyloPhlAn2</a>
to the 23 genomes. PhyloPhlAn works on a set of conserved core genes, so
first we set up the database of core genes, based on <em>Actinomyces
odontolyticus</em> which we picked as a representative since
phylophlan2_setup_database prefers to work exclusively with a single
species identifier:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>phylophlan2_setup_database.py -g s__Actinomyces_odontolyticus -o phylophlan_Act --verbose | tee phylophlan_Act_setup.log
</code></pre></div></div>

<p>Once complete we ran the actual phylophlan2 analysis, which generated
the tree based on the identified concatenated core gene sequences. Of
note, we specified <code class="highlighter-rouge">--min_num_entries 23</code> to force it to only use core
genes found in all genomes, which we deemed important as we had set up
the core database based on <em>A. odontolyticus</em> but we are including other
species as well. Other parameters were chosen based on the PhyloPhlAn
wiki (<code class="highlighter-rouge">--diversity medium</code> due to being a genus-level analysis and
default trimming/thresholding values)</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>phylophlan2.py -i genomes -o phylophlan_Act -d s__Actinomyces_odontolyticus \
--trim greedy --not_variant_threshold 0.99 --remove_fragmentary_entries \
--fragmentary_threshold 0.67 --min_num_entries 23 -t a -f isolates_config.cfg \
--diversity medium --force_nucleotides --nproc 8 --verbose | tee phylophlan_Act_output.log
</code></pre></div></div>

<h2 id="comparing-phylophlan-with-the-pangenome-dendrogram">Comparing phylophlan with the pangenome dendrogram</h2>

<p>To directly compare the agreement between the phylogenomic tree and the
pangenome’s arrangement of genomes based on gene cluster content (and
indirectly, the correlation between phylogeny and genome content) we
plotted them against each other in R.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">dendextend</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">phytools</span><span class="p">)</span><span class="w">

</span><span class="n">panTree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.tree</span><span class="p">(</span><span class="s2">"Actinomyces-23-frequencies.nwk"</span><span class="p">)</span><span class="w">  </span><span class="c1"># the dendrogram from pangenome obtained with anvi-export-misc-data -t layer_orders</span><span class="w">
</span><span class="n">phlanTree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.tree</span><span class="p">(</span><span class="s2">"phylophlan_Act/RAxML_bestTree.genomes_refined.tre"</span><span class="p">)</span><span class="w"> </span><span class="c1"># pylophlan tree</span><span class="w">

</span><span class="c1"># correct the tip names</span><span class="w">
</span><span class="n">phlanTree</span><span class="o">$</span><span class="n">tip.label</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">matcher</span><span class="p">[</span><span class="n">phlanTree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">,</span><span class="m">1</span><span class="p">]</span><span class="w">

</span><span class="c1"># root phylophlan tree roughly, to place the susceptible group sister to resistant group</span><span class="w">
</span><span class="n">phlanTree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">root</span><span class="p">(</span><span class="n">phlanTree</span><span class="p">,</span><span class="w"> </span><span class="n">outgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">panTree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">],</span><span class="w"> </span><span class="n">resolve.root</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="w">

</span><span class="c1"># make a named list of colors correponding to TM7x hosting ability</span><span class="w">
</span><span class="n">colList</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s1">'#a82828'</span><span class="p">,</span><span class="m">10</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s1">'#77398f'</span><span class="p">,</span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s1">'#273075'</span><span class="p">,</span><span class="m">10</span><span class="p">)),</span><span class="n">panorder</span><span class="p">)</span><span class="w">

</span><span class="c1"># plot it</span><span class="w">
</span><span class="n">plot.cophylo</span><span class="p">(</span><span class="n">cophylo</span><span class="p">(</span><span class="n">phlanTree</span><span class="p">,</span><span class="w"> </span><span class="n">panTree</span><span class="p">,</span><span class="w"> </span><span class="n">assoc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="n">rev</span><span class="p">(</span><span class="n">panorder</span><span class="p">),</span><span class="w"> </span><span class="n">rev</span><span class="p">(</span><span class="n">panorder</span><span class="p">)),</span><span class="w"> </span><span class="n">rotate</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">),</span><span class="w"> </span><span class="n">fsize</span><span class="o">=</span><span class="m">0.4</span><span class="p">,</span><span class="w"> </span><span class="n">link.col</span><span class="o">=</span><span class="n">colList</span><span class="p">,</span><span class="w"> </span><span class="n">link.lty</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">link.lwd</span><span class="o">=</span><span class="m">3</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="Act23_phylo_files/figure-markdown_github/unnamed-chunk-19-1.png" alt="" /></p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Rotating nodes to optimize matching...
Done.
</code></pre></div></div>

<p>And they agree exceedingly well.</p>

        
      </section>

      <footer class="page__meta">
        
        


        
          <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2020-10-16T21:53:26-04:00">October 16, 2020</time></p>
        
      </footer>

      <section class="page__share">
  

  <a href="https://twitter.com/intent/tweet?text=%2AActinomyces%2A+comparative+genomics%20http%3A%2F%2Flocalhost%3A4000%2Fprojects%2FAct23_phylo" class="btn btn--twitter" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Twitter"><i class="fab fa-fw fa-twitter" aria-hidden="true"></i><span> Twitter</span></a>

  <a href="https://www.facebook.com/sharer/sharer.php?u=http%3A%2F%2Flocalhost%3A4000%2Fprojects%2FAct23_phylo" class="btn btn--facebook" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Facebook"><i class="fab fa-fw fa-facebook" aria-hidden="true"></i><span> Facebook</span></a>

  <a href="https://www.linkedin.com/shareArticle?mini=true&url=http%3A%2F%2Flocalhost%3A4000%2Fprojects%2FAct23_phylo" class="btn btn--linkedin" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on LinkedIn"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span> LinkedIn</span></a>
</section>


      
  <nav class="pagination">
    
      <a href="#" class="pagination--pager disabled">Previous</a>
    
    
      <a href="/projects/IdrD" class="pagination--pager" title="IdrD domain-level metagenome surveys
">Next</a>
    
  </nav>

    </div>

    
  </article>

  
  
</div>

    </div>

    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <div class="page__footer-follow">
  <ul class="social-icons">
    

    

    <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2020 Daniel Utter. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>
  <script src="https://kit.fontawesome.com/4eee35f757.js"></script>










  </body>
</html>
