<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.20.1 by Michael Rose
  Copyright 2013-2019 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>IdrD domain-level metagenome surveys - Daniel Utter</title>
<meta name="description" content="From Sirias et al., bioRxiv">


  <meta name="author" content="du">


<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Daniel Utter">
<meta property="og:title" content="IdrD domain-level metagenome surveys">
<meta property="og:url" content="http://localhost:4000/projects/IdrD">


  <meta property="og:description" content="From Sirias et al., bioRxiv">







  <meta property="article:published_time" content="2020-10-18T13:47:50-04:00">






<link rel="canonical" href="http://localhost:4000/projects/IdrD">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": null,
      "url": "http://localhost:4000/"
    
  }
</script>






<!-- end _includes/seo.html -->


<link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Daniel Utter Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css">

<!--[if IE]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--single">
    <nav class="skip-links">
  <h2 class="screen-reader-text">Skip links</h2>
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          Daniel Utter
          
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/About">About</a>
            </li><li class="masthead__menu-item">
              <a href="/CV/CV.md">CV</a>
            </li><li class="masthead__menu-item">
              <a href="/Research">Research</a>
            </li><li class="masthead__menu-item">
              <a href="/Methods">Reproducible Methods</a>
            </li><li class="masthead__menu-item">
              <a href="/Misc">Essential 'omics commands</a>
            </li></ul>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      



<div id="main" role="main">
  


  <article class="page" itemscope itemtype="https://schema.org/CreativeWork">
    <meta itemprop="headline" content="IdrD domain-level metagenome surveys">
    <meta itemprop="description" content="From Sirias et al., bioRxiv">
    <meta itemprop="datePublished" content="2020-10-18T13:47:50-04:00">
    

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline">IdrD domain-level metagenome surveys
</h1>
          


        </header>
      

      <section class="page__content" itemprop="text">
        
          <aside class="sidebar__right sticky">
            <nav class="toc">
              <header><h4 class="nav__title"><i class="fas fa-file-alt"></i> On this page</h4></header>
              <ul class="toc__menu">
  <li><a href="#overview">Overview</a></li>
  <li><a href="#the-process">The process</a>
    <ul>
      <li><a href="#identifying-metagenomes-likely-to-contain-populations-with-idrd">Identifying metagenomes likely to contain populations with idrD</a></li>
      <li><a href="#preparing-the-metagenomes">Preparing the metagenomes</a>
        <ul>
          <li><a href="#downloading-and-sorting">Downloading and sorting</a></li>
          <li><a href="#mapping">Mapping</a></li>
          <li><a href="#importing-the-results-into-anvio">Importing the results into Anvi’o</a></li>
          <li><a href="#retrieving-the-coverage-and-variability-data">Retrieving the coverage and variability data</a></li>
        </ul>
      </li>
      <li><a href="#inspecting-metagenome-coverage">Inspecting metagenome coverage</a>
        <ul>
          <li><a href="#preliminary-data-processing">Preliminary data processing</a></li>
          <li><a href="#filtering-metagenomes-based-on-category">Filtering metagenomes based on category</a></li>
          <li><a href="#binning-metagenomes-by-origin">Binning metagenomes by origin</a></li>
          <li><a href="#per-nucleotide-coverage-to-visualize-region-specific-coverage">Per-nucleotide coverage to visualize region-specific coverage</a></li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

            </nav>
          </aside>
        
        <h2 id="overview">Overview</h2>

<p>This is a reproducible, narrative description of the methods used to
investigate the representation of <em>idrD</em> in real-world environments,
specifically the human microbiome, as reported in <a href="https://www.biorxiv.org/content/10.1101/2020.02.20.956912v1.full">our
manuscript</a>.
The motivation for mining metagenomes came from the biochemical
characterization of IdrD, which represents a novel contact-dependent
toxin involved in self/non-self identification between bacterial
strains. IdrD is particularly interesting, structurally, it contains
multiple subregions with varying nucleotide-level conservation. The 5’
end of the gene (N-terminus) seems to behave “ordinarily” with expected
nucleotide-level divergence between taxa; the middle usually has a
relatively conserved domain for recombination (e.g., Rhs); and the
C-terminal domain (CTD) is much more variable. This CTD contains the
active site for DNase activity. So, a priori we expect some lumpy
coverage reflecting the different domains, as they represent different
levels of sequence conservation (i.e., frequency in the pool of short
reads that is a metagenome). Thus, we were interested in investigating
its presence and diversity in complex, host-associated microbiomes to
understand the CTD’s prevalence, diversity, and population specificity.</p>

<h2 id="the-process">The process</h2>

<p>We started with a FASTA file of full-length idrD nucleotide sequences,
obtained by manual curation of idrD hits from a combination of HMM and
tBLASTn searches of EMBL and NCBI. The sequences can be found here:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">&gt;</span>Acinetobacter_baumannii_strain_XH858_idrD
ATGAACAAGAATAGTTATCGCATTATTTATAGCAAAGCACGTCAAATGTTTGTGGCCGTGGCTGAAAATGTACGAAGCCAAACCAAGACTTCTGGTCAAAGTGAAGCAAGTACCCAGAACAAGATTGATAACACTGAATCTCAGGCATTTCATCAACTCTGGCAAGTGAAAGCATTGGTTGCCAGTATTAGCCTGTGGATGCCTTTAGCACCTGTTTATGCGGGTATTGTAGCGGACAGTGCCGCAAATGCTGCGAATAGAGCAGTTATTGGTGCTGGAAAGAACTCAGCAGGTACAGTCGTTCCTGTCGTTAATATTCAAACTCCAAAAAATGGGATATCCCATAATATTTATAAACAGTTTGATGTATTGGCTGAAGGTGCAGTACTCAATAACAGCCGTCAGGGTGCTACGACAAAAACAGTAGGAAACGTTGCAGCGAACCCATTTTTAGCTACAGGTGAAGCACGAGTTATTTTAAATGAAGTAAATTCATCAGCCGCTTCACGCTTTGAGGGGAACTTGGAAGTTGCAGGGCAAATGGCTGATGTGATTATTGCCAACCCGTCTGGTATTAGTATTAAAGGTGGTGGCTTTATTAATGCCAACAAGGCAATTTTCACCACAGGTAAACCTCAGTTAAATGCTGATGGCAGTATCAAGCAGTTTACGGTTGACCAAGGAAAAATTACGGTCTCAGCGAACCCCAATTCTAAATTTGGTTTGGGTGGAAATAATAACGATGCAAATTATGTGGATCTCTATGCGCGTGCATTAGAGCTTAGTGCTGAGCTACGTGCTAAAAATGATATTCAGGTGATTGCCGGGGCAAATAATGTCAGTGCAGATTTACAGGATGTTACAGCTAAAACAGGTACTGGAACAGCACCAACATTGGCTGTGGATGTTAAAGCATTAGGCGGGATGTATGCCAATAATATTTATTTGATGGGAACTGAAAAAGGCTTAGGCGTGACGAACGCAGGGACTATTCAGGCAGTCAACAATCTGGTAATTACCAGTGCCGGAAAGATTGAACACAGTGGTACAATCAGTTCGACCAGTAAAACTCAAGGTTTAGTAAACATACAAACAACGGGTACGGGTGCAGCTTCCGATATCAACTCATCAGGTAGTATTAATAGTAATAGCATGTTAAATATCGATTCAGGTAATAACCTGAATGTAAATGCTAAAGAAATTATTATTAATAACAGCAGTCTAGCATCTTCACCTTTAATCGTTAATACCAAAGGTAATATCAATTTAGCAGCAAACACCCGCATTATGGATGATTCACAAGGTGGAGATGTCTATATTGATGCAGCAAATATCAATCTTGCCGCTGGATCGGAACTCAAAAGTAACCGTGGTACAGCAACCATACAAGTACAAAAAGACTTAGTTGCAGCCAAAGGCGCAAAACTGATTGCTGCCCAAGACTTAAATGTTTTAAGTAATGGTAAGTTATCACTTACTGAGAACCATATACAGGCATCATTAGGTACAATCAACCTACAAGCAAATTCGGCTAATACCCAAAATCTGATTGACCTTCAAGGTGGAACGATTTATGCGGGTAAAGATTTAAATTTATATAGCACTGATAATATTAATATGAAAAACTTAAGTTTTTCAATAGAAAATAATAAAAATCGAGTAAAAAATATTAATATTTATAGTGGTGGTAGTTTATATTGGGATAATACTGCTTTTTTAATACCTGACCTAACTGGAAAAATTTCACTGAGAGCGGAAAAAGATTTAACAATATTGGGTCAAGGTGGTTTAAGTGCAGATAACGGTATTAGTTTATATGGTGGAAATGTAAATGTTTCTTCACTTAATTTAAATTCAAATATTGGTAATACATTGGTTTTTGCAAATAATAATATTGATATTAAGGGGGTAAATATAAAGAGTAATAATGGAAATTTAAATTTTTCAGCAGAAAATGGCTATATCAAGATTGCTGATTCGAATTTGTCAAGCGAAAAAAATAAATTATTTTTAATTTCTAAAGATAAGCAAATTTTTAATAATACTAGTTTACTTTCAGGAGGGGATTTAATATTAAATACTATAAGTAATAAAGGTGATGTAGAAATTAATCAAGATGGAGGGTTATTTTCTTCACCAATTAAATCAAATAATGGTAATGTAATTATTGATTCTAAAGGGATGTTAAATATAACAACAAAAGACTTAAATAAATTTGTTGATATCAGTGGTGAGAAAATAGAAATTTCTTCTTCTAATAATATGAAATTAGATGGTATTAATGTATATTCAAATGGTAATTTATCTCTACATGCAGATAAAGATATATACTTGAATACAGTTCTGCCAAACTCATGGACAGGGGGTGGTTTTGGTAGTGAAATCATTAGAAGCCAAAAGCATATATCTATCAGTAGTAAAGAGGGCAGTGTTAAAATAGGATTAGAAAACCAAGCTTTAAATGCTCCTGTCGTTGGCCTAAAAGCTAATGGCGTAATTTCAATAAAAAGTAAGTTAGATCAAAGTTATCAAAATACCTCGCTAAATGCGGGTGCTATTACGCTTCACTCAGATGCCGGAAGTATTAATAATAGTAAAAATTTTAGAGCAATTGCTAATACAACATTTTTTCTCGAAAGTGATAGCGAATTAAAAAAAATAAATGGTAACTTAAGTTTATATGCGAAAAAAGATATTCAACTAGACGCGATTAACTCTCAAGGACGAGCTTTTGATCCAAATTACACGCGTATGATTAAACCCGTGCTCTCAAGCCAAGGTTTAATGGATATTCGTGCTGATGGAGTGTTTCAGGTGAATGGTAAAGCACCTCCAGCACAAACTGCGATGTGGGATCAGTTGAAAATGGATAGAGCATATTTTACTTCAAATGATGGTATTAATATTCTAGCAGGAACTGTCAAGATGTATGCTGGTGATCTAAAAAACACATCAAATACTGCACCAATCAATATAATATCAACAGGAGATATCGTTCTTGATGCCATGAACTATGATGTAAATATAGGGCAGTTAGATCTGATGTTACCAGCACAAAAGCTTCGAGATGAATTAGAAGCTAGCGGTAAAAAGGATTCTGTTACATTAAATACAATCAACCAGCTCAATGAAGAAATTGCATTTTATTTATCAAGATCATTAAATGGAACAAGAAGCCAAAGTAGCCATATTGATTCAGCACAAAATATTAATCTCCTATCTAAGAAGGGAATACTCATTAGAGCTTCAGAATTATATGCTAAAAATGAAGTAAATATTGAAGCACAAGGGTTATTAGCACGTGATCTTACAGGGGCTCAGGTAACTGATTATATCGATACTTCAATTTTAGTAGATGGGGTACATGATATTTATAAAAATGGTGAAGCTACTAATTCAAATTATGAAGAACGCTCTGATTTTCATAATTCAATTGTTAGTGGTGATAATGGAATCAATATTAAATCAACAGGTTCAGCCAAATTAAATTGGTATGATAATCCAAACCTATTATCTAATTACGCTACAGTCACTTCAGCAAAACCTATACAGTTAAAAAAGAAAATAAATGATGGAAAAAATAATATTGTTTTTAATGGTGCAGAATTAATTTCATCAGGAGGAGATGTCAATATTCAATCTAATGCAAATATTGTATTAGAATCAAGTCAAGATGAGATTTATAAGCATAGTACCAGTCAAGTTATAAAAAAAACTTGGTATGGAAAAAAGAAATCTACAGAAACCTATACACAGAGCCAATATGGGACAGCTGTTCCAACAACAATTTTAGCAAAAAATATTAGTCTGAAGGCAGCAGGAGATATAGATTTCTATTCAACGTTGATGAAAGCAAATAGTGGTAATGTTAATATTACTGGCGAAAATCTTTACTTTCTAACTTCAAATAACTATGCAAATTCAACTAGTACAACGAAAAAAAACTCATCAATTTTAGGAATTCCACTTAATAAATCTAAAACGACTAGTAGTCGTAGTCAAATTTCGCAACTTCCAGTAAAGCTGGTTGGAGATTATCTGTCAACTGAATCGAAAAATGATACAGTTTTTGTTGGGACAGAGTTTGATTATATGAAAAATGCGTCAGTTAAGGCAGGTGGGCAGATTGCTTTCCTTGGGGCATCAGAAACAATAAGTCAAGCTTTAAAAGTTGAAAAGAGCAATATTGTTTGGCAATCAATGCAAGATAAAGGCTCTATTACTGAAACTACTAAACTCCCAAATTTTAATGGACCTACTCCACCTACTTTTAAAGCATCAGGTGGCTTAATAGTACAGGTGCCTATTATATCTAATAAAAATAATGATATCCGTACAGAAGTGATTAACCTTGCTAATCAACCAGGTAATGAATATCTCAAAGGTTTTATTGCCAGAAAAGATGTAAATTGGGAGGCGGTTAAACTTGCACAAGAGAATTGGGATTATAAATCACAAGGTTTAACTGCTGCTGGTGCTGCAATTATTGTAATCATAGTTACAATTGCTACGATGGGTTCTGGTACTGCAGCTGCAGCTGGTGCTGCAGGTGGAACTGCTGCTAGTGGGACTACAGTAGGTCTTGGAGCAAGTATGATTGGAACTGCTGGTGTAACTACGACTGCAGCAGGAGCAATTGTTCCATCCACTTTAGGTGTTATGGCTAATGCAGCTGTTACAAGTTTAGCAACTCAAGCAAGTATTAGCTTAATTAATAATGGAGGTGATATTAGTAAGACACTTAAAGATTTAGGAAGTAAAGAAACTGTTAAAAATTTAGCAACTTCGGTTGTTACTGCAGGAGCCTTGAGTACAGTTGGTGGGCTTGATTGGATGGAAAAATTAAAAGATATGGGTACAGATCCGAATCTTTTAGTAAATCTAGCAGGGCGTACTGGTGGAGCTATAGTTGAAGCCGGTCTAAGTGCAGGAATTAGTAGTAGTATCAATGGTGGAAGTTTAACTGATAATTTACAAGGTATGTTACTTTCTAGTTTCGCAACAGCTTTACAGGGTAGTTTAGCGTCAAAAATTGGTAAAGAACTAGGAGGAATGAATAAAACAGATTTCGAAAATATTTTACATAAAATAGCCCATCTTGCTGCTGGGTGCGTAGCTGGTGCTATACAAAAACAATGTGAATCTGGCGCTATAGGTGCAGGAATTGGAGAGTTTGTCGCAGAGTTCATGCCAGCAGCCAGTGGCAATGGGGTGTATAGTGATGACGAAAAAGAAAGAGTTTTAGCAGTTGGTAAAATTTTGGCAGCTGGTGTAGCTGGTATTACTGGGTACGATGTAAACGTTGCATCAAGCTCTGCAAATACAGCACTGTTGAATAATGCATTATCAAAAACGGATGTAGATGTATTAATTCAAAACTTAAAAAGCTGCAAAAATGAATCACAATGTCGAAATGAGATTCAAAAAGCAGTTCTCAAGAGTAATTATAATCAAAGTAATGCAGTGAATTGTCTTTCGTTACTTTGCTCAGATAATCTTGAGCAGGCAATTAGTGGTTATAATCAGCTTAGAGGATTTATAGCAAGTGGTAAATTTTCAGGAGATAATTTAAAGCTATTAAATAATTTACTCACTCAAAGTACTAAAGATATTACTCTTCAAATTAATAGACTACCTCAAGTTCAAGCTTATCTAAATTCCCCTTGGACACCGCCAACAAAGCAACAGATGGCACAGATAGTTATTGATATTTTACCAATGACTAGTGATCCTTCTGCAATTTATTCAATTATTACAGGTAAAGCAGTTGTTACACAAGAAGAGGTAAGTCGTTTTTATGCAGCTGTTGGTTTAATTTTGCCAATTACAATTGGTCGTTCTTTAACAAAAGCCGAAATTGCTGGTTTGAAACAGTCTTCGCAAAATCCACTAAGCGGTACAGTAAGTATAAAAGATCTCCCAGCTCTACCAAAATCTGCACAAAACTTCGTTATTGAACAAACAGCACAATCTAAATATACAAAACTTTTATATGAAGCAAAGAATGATCCTAGTTTAGGAACTTGGAGTACTTCTAAAATAGGTCAAAAGGGTGAGGAAATAGCTGTACAGTTGGTTAAGGGGAGTGGTTTTACAGATGTTATAAGTATACAAAACGCTAGTGGTAATGGTATTGATATTATTGCAAAAAATCCTCAGGGAAATTATGTTTACTTAGAAGTAAAAACATCTGCTGTTGGAAAGATTGGTAGTTTATCTAAAAGGCAAGAAGATATGAACTTCTTTGTAAATGATATATTAACAAATGCCGCTGCTAAAACTGGGCGTTATGTAAACATAAGTGATACTGTAAATGCACAAGCTAAAATTATGTTAACAAAATATAAGCAACAAAAAGCTGTAGGTAATATCTATGGAGCAGCTATAGGGGTAGATTTAAAAAATGCTCAAATCTTAATCTCTTCTTGGACAAATAAATAA
<span class="o">&gt;</span>Cronobacter_turicensis_z3032_idrD
GTGTCTGAACCGTTAGCTGCCCGTTATCAGGACCCGCTCATCCACAGCTCTCTGCTCGCCGATGTCGTCAGCGGTGTGGTCGAAGGCGCCATTTGTCTTGCTGCGTTTACGGCGGGTGCGGCGATGATGACAACGGGGCTTGGCACGGTCGCAGGCGTTGTACTGATTGCCGCTGTTTTCGGCAGCGGCATCGCGGAAGAGGCAGGCGATCTGGTCGGCCAGGGTGTTGATGCCGTGCTGGATTTCTTCGGCTGGCGTGGGCCACCGGATGCCTTTATTACCAGTGGCTCGCACAACGTTCATATCATGGATCTCCCTGCCGCCCGTGCGGCGGGTACGGTCGATCATGATTATCTCAATACGCCTATCCCGGCAGAGAGCTTTGCCGATAAGGCGAAGGCGTTCGCGATAAACACGGCGGTTACCATTCTCGAAGTGGCGCAGTTTACGCTGCACCCGTTTGATAATCTGGCTGCGGGAGCCGACGCGATTGCGAGCAGCGGCTGGCAGGGCGTGAAAAATTTCGCGGGCAGCGTCTGGGATAACTTAACCCAGCCGGTCGTGGCGGGGGCCAGCCCTTTTGCTAAAGAAGCGCCGCTGGATACTGTCGAGTGCACCAAAGGCCATACGATCACCGGCGGTAACTTCCTGGCGGAAGGTTCGAAAAAAGTGCTGATCAACGGTCAGCCCGCGTGCCGTGATGGCGATCGCAGCACCTGCGAAGCAAAGATAAAAGTTAAAGAAAACACGCGTGTGCGCATCGGCGGCGAAAGCATTGTGGTGCGCGATATCCGCAGCGGTAAAAATTTCTGGGCGCGGCTTATCGGTAACGCCATCGGCAGCCTTGGACCGGGAATATTGCGCAATCTCAGTAAAGGGATGTTAAAGGCGATATTCAGCCGCCAGATCCTGAAAACCTTTTGCTGTCAGCTCGCCGCCGATCTCGGCATGGGATTAACCACGTTCGGGCTGATTCAGGCGGGGAAAGTGGGCAGCGAAGCGCGCCACACGCAGCACCCGGTGGATATCGCCAGCGGCGCGAAAATCCTCGCGGGCGGCGAAGACCGCGATTTCACGCTGGAAGACCGCATTCCGCTCATCTGGCAGCGTGTTTACAACAGCCGCAATCTGGCGACCGGTATGCTCGGTACCGGCTGGCTGCTGCCGTTTGAGACCCGCTTCTTCCGTCTTGAGAACAATCAGTTTATCTGGCGCGATATGTCGGGCCGCGATCTGGGCTTCGGCGAGCTGACCCCCGGCGACGTGGTGGATTATCTCGAAGACGGCGTCACGCTCTATTACACCGTCAGCGGCACGCTGATGCTCCAGATGACGAGCGGCGAATATCACGTCTACGAACCGGACCCGACGCGTCCGGGCGAGTGGCGGCTGTTCCGCATCTACGATCGTCACGAAAACTGCCAGTATTACAGCTGGGATGAACACGGCAGGCTGGCGCGCATTTCCGGCGACAACGAAGCGCTGGATGTCGAACTTGCCTACGAGAAATCGCATGGCCGTCTCGCCAGCGTGCACCAGGTATGCAACGGCGAGCGCCGCCTGCTGGTGACTTATGGCTATAACGAAAACGGCCAGTTGATCGAGGTCACCGACGCGGACGGCATCGTCACGCGCCGCTTCGGCTGGGATCGCGCCAGCGACATGATGGGCTGGCACAGCTACTCCACCAACCTCAGCGTGCATTATCAGTGGCAGCCCGCCGCCGATGCGCCCAACTGGCGCGTGTGCAGTTATCAGGTGCTGGACGACCAGGACAACGTGCTGGAGCGCTGGCGCATCGACGCCGACGAGGCGAAACGCTGCGCCACGGTAAGCTGCGACGCGGGTTTCTCGACGCGCCACTGCTGGGATTTCCTCTACCGCATCACGGAATACACCGACCGCCACGGCGGCGTGTGGCGCTACGAGTGGGCAGATTATGCCGAGCTGCTGAAAGCGGCCACCACGCCGGACGGCAGCCGCTGGGTGTATGGCTACGACGAGCACGGCAACCTGACGGAAGTGCGCGATCCGCTCGGCAACAGCACATTCACCACGTGGCACCCGGTGTTCGCGTTCCCGGTGAAGGAAGTGCTGCCGGACGGCGCCGCCTGGCAGTATGAATACAATGCGCGCGGCGATGTGGTGGCGCTCACCGACCCGAAAGGCGGCGTGACGCGCTTTGAGTGGAACGAGCAGGGCGACCTGGTTCAGCAGACCGACGCGCTGAACAACACGCACCGGTTCTGGTGGAACGAGCGCGGACAGCTGGTGCGCGACGAGGACTGCTCCGGCAACCAGAGCCAGCGGCTTTACGACGACGCGGGGCGGCCGCTCAGCGCCAGCGACGCCGAAGGCAACACCGACCGCTGGGCGCTGACCGCGGCGGGGCGTCTGCGCACCTGGCGACGGGCGGACGGCCGCGAGACGCACTACGAATATGACAACGCCGGGCTGCTCTGCGGGCAGGACGACGACGGGCTGCGCGAGCGCAAGGTAACGCGCAACGCGCGCGGGCAGGTGGTGAGCGCCGCCGACCCGGCAGGCCATCTGACGCGCCTGCGCTATGACCGTTTTGGCCGTCTGACGACGCTGGTGAACCCGAACCGCGAAAGCTGGCGGTTTGAGTATGACGCCGCGGGCCGCCTGACGGGCCAGCGCGACTACGCGGGCCGCCTGACGGAATACCGCCACGACGCGCTGGGCCAGGTGACGGAGGTGATCCGCCATCCGCTGCCGGGCAGCCCGGACGCGCCGCTGGTCACCGCGTTTGAATATGACGTGCTGGGCCGCCTGACTGCGCGGGAAACCGCAGAGCACCGCACCGAATACCGCCACGACACGCTGTCGCTGGAAATCCGCCGCGCCACGCGCGCCGAATGGCGGCAGGCGCTGCTGGAAGAGCGCGAACCGGCGTGGGACGCCGTGCTGGTCTTCACCCGCAATGCCGCGGGCGAACTGGTGAGTGAGGAGAACCACGGCGGGAAGTTTGAGTATGAGTACGACGCGCTGGGCAACCTCAGCAGTACAACGTTTCCGGATGGCCGGGAGCTCGCCACCCTGCGCTACGGCACCGGCCACCTGCTGGAGATGCAGCTGCGCCACGGCGGGGCCACGCACACGCTGGCGGCATACGGCCGCGACCGCCTGCACCGGGAAATATCCCGCAGCCAGGGTGTGCTCTCACAGGAGACCCGTTACGATTCCGCAGGCCGCGTCACCCAGCGCACGGTGCTGGATGCGCGACGCGAACTGGTGTTCGAGCGTCGCTACCGCTGGGACCGCACCGACCAGATAGTCCAGCAGATACACACTGACTCAACCCCTGCCACGCCGGGCGAGAAATACAGCCAGTACCTGTGGGGCTACGATGCCGCGGGCCAGGTCACAAAAGCTGTCGAACCGCAGAAAGAAGAGCGCTTCTTCTGGGACGCCGCGGGCAACCGCACTGAGGAGCACCGCAACCCGGTGTGGCACAACCTGCTGCTGCGCCTCGACGGACTGAAGCTGGATTACGACGGGTTCGGGCGACTCACGCGCCGTCAGGATAAAAGCGGCGTAGTGCAGCATTTTGCCTACGACGATGAGCAGCGGGTAAAAGAGATTCGGTTTGAGGGGAACACGAAGTTCCGCCGGGTGGAATACCGCTACGACCCGCTGGGGCGACGCACGCATAAAATCCTGTGGCGCTACGGAGAGAAAGACCCGGAAACCATCCGCTTCGACTGGCAGGGTTTACAGCTTGCAGGTGAACAGAGCGACCGGGAGCCTGACCACTACGTTCAGTACGTTTACACCGAAGGCAGCTACGAACCGCTGGCGCGGGTCGACAGTGTGTATGACGACTGTGAGATTTACTGGTACCACACCGAACTCAACGGCCTGCCGGAGCGGGTTACCGACGCGAACGGCCAGACCGTCTGGCGCGGACAGTTCAGCACCTGGGGCGAAACCGAACGCGAGCTCAGCGTGCCGCAGTGGCAGGTGCCGCAGAACTTACGCTTCCAGGGCCAGTACCTCGACCGTGAAAGCGGGCTGCATTACAACCTCTTCCGCTATTACGACCCGGTTGCCGGGCGTTACACCCAGATGGACCCGATTGGGCTTCTTGGTGGGATTAATACTTATGGATATGTGCCGGACCCATTAACTTGGGTAGATCCGCTTGGTCTGTGCTTCAGTAAGCGCTTGGGTGATTTTGGTGAATCCCATGTTAAATCTCAGCTCGAACGCTCAGGAAATTATGCAGATGTTTTTTCAGTGCAAAATAAATCTAATAACGGTATTGATTTAGTCGGCTTAAGACATGACGGGAAATATGACTTCTTTGAAGTAAAAACAAATACAACCGGATCTGTCGGCAAATTAAGTGATCGGCAGCTAAGTTCAAAAGATTTTATTGAAACTGTTCTTGGAACAGATGTCCGGAAAGGAGGCTATGACCTTAACGGGCACAGAGTCAATGATATGTTGGATAATATAGGTGATACTTATGTTGTCGATGTTTTCGTTGAAAGAGCCTCTAATGGGCGTTGGTTTGTTTCAAAGTATTTGATGAGTATTTGGGAGTAA
<span class="o">&gt;</span>Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD
ATGGTGTGGGCGCGCGACGTGCTTGGGAACGTGACGGCCTACGCCTACGACGACGCCGACAACCTGATACACATTGAGGAAGCGGACGGCAACGTGCACAACCTCGTGTACGACGCCTCGGGTAACCTCGTCAAGGCGTCGGACCTGCTGCACGACGTGGAGTTCGACTACGGCCCGCTGGGCACGCTCACGGGCAGGCGGCAGTTCGGCCGCCACGTGCGTTTCGGCTACGACAGCGAACTGCAGTTGCGCGAGATTATCAACGAGGGCGGTGAATGCTATTCCTTCCGTCTTGACGGACTCGGTCGTGTCGTGGAGGAGACTGGCTTTGACGGCCTGCAGCGAAAGTACCTTCGTGATGGTGCTGGCAGGGTGATACGGGTTGTACGTCCCGAGGGAAGACAGACAGAATATGAGTATGATGGTGTGGGCAATATTCTGCAGGAAACACAGCATGATGGTCGAACAAGCCGCTTTGCTTATGACGGGGACGGACAGCTGCTGCGTGCTGAGAACAAGGAGATGAAAGTAGCCTTCAAGCATGACCTTGGGGGACGGATAGTGAAGGAGACGCAGGGCGGGCACTGCATCATGCGCAAGTATGACCGTACAGGACGCCATGTGCACACCGAGAGCAGTCTTGGTGCAAGTATTGACTATTCCCACGACAAGGATGGCAACCTCAGTGAGATGAACAGCGGAGGATGGTCAGCCAGATGGCAGCGTGACAGGGTAGGACTGGAGACGGAGCGTGAACTTACAGGCGGAGTGCATGTAAAGACACATCATGACAGCCTTGGACGTGAGACCTACAAGTCAGTGGGCGCACGCAATGTCGAGCAGTTCCGCCGCAGCTACACCTGGGGCATCGGCAACCGTCTTCATGCAACCGAGGACGGTCTTACGGGGCGCCGGGTACGCTATGACTATGATGAGTTCGACAATCTCCTTTCTGCAGAGTACAAGCAGGGGAATGACGTTGAGCGCATCTACCGTATTCCTGACCGTATCGGCAATCTCTTTGAGACACGTGAGAAGGATGACCGTAAGTACGGTGCGGGCAGCCGTCTGACGGAAGACCGTGACTACTTCTATCACTATGACTGCGAGGGCAACCTGGTGTTCAAGGAGTTCAGGACGCTGGACTGGGCCGGTGTGTCCGTTCCGCTTGGCGGACAGAAGGCGCACCTGGAGGAGGAGCTCGGCATAACCTTCCGTGCCTTTGGCGCAGGCTGGCGTTATGACTGGCAGAGTGATGGTATGCTGTCGAGGGTTGTTCGCCCCGACAGCAAGGAAGTGTCATTCGCATACGATGCGCTGGGCAGGCGGACCGAGAAGACTTATGAGGGCGTCGCTACTCATTTCGTGTGGAACGGCAATGTACCGCTGCACGAGTGGCAGGAAGTTGCAGCTGATGCCGGAAAGGCAGATGTCACCACCTGGCTTTTCGAGCAGAACACATTCATTCCTGCTGCCAAGCTCGCAGCCAACGGCGAGAGTTTCTCCATCGTGTCTGACTACCTCGGTACTCCCCTGCAGGCATTTGACAATAACGGCAACAAGGTTTGGGAGCAGGAGCTTGACATCTTCGGCAGAAAGAGAAGAAAGAACAACAATAATTCCTCCTTCATTCCGTTTAAGTATCAGGGGCAATATGAGGATGTTGAGACGGGGCTGTATTACAACCGCTTCCGCTATTATGAGCCGAATACAGGAACCTATATCAGCCAGGATCCGATAGGGTTGGTGGGCGGAAACCCGACACTATATGCGTATGTCGGAAGTCCTAACAACTGGTATGATATATTTGGCTTACGACCATTCGGACATGCTGTGGGAGATATTGGTGAGAAAGCTGTAATTAATGACCTTAAAAAGAACAATTATGAAATTATAGACGTAAAATACGGCTCTAACAATGGCATTGATGTGCTTGCCAAGAATCCAAGTACAGGAAAATATGATGCCTTTGAGGTAAAATCCTCTACGGTAGGAAAATTCAATCTTTCTGAAGCTCAGATGGATCCTGACAACTTTGTAAGGACACGATTGGAAAGAGCTGAATCACATGGTATGATTACCGAAGATGCTCGAATAGATATTATGTCTAACTTGGGAGAAAGAAAGGTTGCTTATGTGGATATAAAACGAAGAGATAATGGGGCTCTGTATGCTGAATCAATCCATTATGAGAGTTGGGATGCAGAGGCAAAAAGAATAGAAAAACTAAAAAAAGGAGGACATCATTAA
<span class="o">&gt;</span>Proteus_mirabilis_BB2000_idrD
ATGGGAAATGAATTTTTCGCAGCCAAGCAAGGCGATAAATTAATGCACTCGTCAATCTTCGCCGATATCGTGTGTGGTGTGGTTAAAGGGGCGGTTTATGCCGCGATAGGGACGGCCGCTGTGGCGTTGACTATTGGTACAGGTGGTATGGCAACGGCCGCCGGTGTTGCCATTTTTGCCGGTGCAGGGCTGGCGGCGGGCTTTACCATCGGTGGATTGATTGATGATGCCGCTGATTGGGTCGCGGATGGGATCGATAGCTTGTTCGGGCTAAGTAACCCTGATGGTGAAATCGATACTGGCTCGAAAAATGTACGTGTCAAAGGCAAGGGAGCCGCTCGCGCAGCGGGAAAATTACCGTCACCCGTACTATTAGCCTCAATTGCGGCAGATAATACCCCACCGAAAGAAGAAAAAGGCACCGAGATTGCCTTACGCGTATTGAAAAATACCGCCATGATGGCGACTGGGGCATTTTCTTTATTGGCTGGTAAGGTGCTAGGGAGCCAAACCAACAATGCGGCACCCATAGCGCTCGATCCCTCCGAGTTTCCCGAATACCCCTCACCGGAGCAAGGTTTTTTTGACAGTATTCTTTCCCCAGTAAAAGCTGCGAAACACCCGAATGCTGAACCTATGCCTTTGGATACCATCACGTGCGATAAATGGCATTGTGCTAGCGCCCCTTACTTTTTAGCCGAAGGGTCGAAAAAAGTCCAAATTAATAGCCAACCGGCGTGTCGTAACGGTGATAGAAGCACTTGTGAAGCCAAAATCAGTGATACGCAAGAAAAAGGGATTGTGCGTATTGGCGGTGGCAGTGTGGTGGTACGCGATATTATGAGTGGGCGTAACCCCTTTGCCGAAATGTTTGGTGAAATTCTCGGTGGCCTTGCTGTTGGAGCGGCGTCTACCCTGTTTCGCAGAGGGGGACTCAAGGCGATTAAACAGTGTTTGAATAGTGTAGGCTGTACTTTGGTGAGTGAAATCGCCAGTACCGCGGTGGGCAGTATGGTAGTCGCCTCGGTGACCCAAGCCTTTAATAGTGTGCGTCACCCCGTGCATGCGGCTACGGGGGCGAAAGTCCTAAATGGAGAGGATGATACCGATTTTATTATTGATGGTGTGTACCCGTTAGTCTGGTCGCGTATTTACCAAAGTCGTAATACGGGTGAAAATCGTTTAGGACGCGGTTGGTCGATGCCGTTTGATGTCTTTTTGACAATTGACGACACAGGTAAAGGGCTTGAGAATGAAAACATTTATTATCATGACATGTCGGGGAGGCGTTTAGCGTTAGGAAAAATCGCCCTTGGACAGAAAGTGTTCTATCAAGATGAAGGCTTTACCGTCTATCGCACGACAAACAATCTTTTCCTTGTTGAATCGGCGGAAGGTGATTATCAACTTTTCGAAGCCAATCCACATAAAATCAATACACTGCGCTTAATGAAAAGTGCTGATCGTCATAATAACGCTTTACACTACCGCTATGCTAATGACGGTGAATTGGTACAAATTCATGATGATGCGTATCTGACCGATATCCGGTTACATTATGATGAAATCACCCAACGCTTACAGTCGGTGACCCGCCACCAAGGACAAGAAGAAAAAACGCTGGTCACTTATACTTATGATGCACAACAGCGTTTAGTGCAAGTCACTAATGCGGATAAGCGAGTGACTCGTCGTTTTGGCTGGGATGATGAATCGGGTTTAATGGCCATGCACCAATATGCCACAGGGGTCAGTAGTCATTATCGTTGGCAACGTTTTGATGCCTTTACCATTGAAGACAATGAACCTGAGTGGCGAGTGGTTGAACACTGGCTTAAAGACGGTAAACGCTGTTTAGAGCATACCGAACTGACGTATGATTTAGCTCAACGAACCCTCACCACCGTGGAAACGGGGGGTGAAACCACCTTCCGTCGCTGGAATGAACAACAGCAAATTATCGAATACACCAATGCATTGAATGAAACGTGGTGGTTTGAATGGGATACCAGCCGCTTATTAACGAAAGCAATAGCCCCAGATGGCAGTGAATGGGGGTATACCTATGATGAGCGCGGTAATTTAACCCAATCGACCGATCCGGAACAACAATCGACTTGCTATGATTGGGACAAAGATTTTGCGTTTCCCACGGCACAAACTTTGCCAAATGGGGCGGCTTGGCATTGGGAGTACAATGAGCACGGCGATATTCGTCGTGTGATTGACCCGCTTGGCCATATCACGCGCTTGGCGTGGGATGACCAAGGCCTGTGTCTGGGGCAAGTGGATGCTAAGGGTAATGAAACCCACTATCGCTATAATGCCCGCGGTCAGTTAATCGAGCAACGGGACTGTTCGGGTTATCCCACCACCTTGACCTACGATGATTGGGGACAACTTCGCTCGTTGACCAATGCACAGAATGAAACCACGACTTACACCTTTAGTGAAGCGGGGTTGTTGTTAACAGAGCGCTTACCGGATGGGACAGAAAACCGTTATGACTATGATGCCACCGGGCAATTAGTGGGAATAACGGATGCCGGAGAGCGCCATATTCTGTTGCGCCGTAACCGTCGTGGACAAGTGATAGCCCGACGCGATCCGGCAGGGCATTGGTTGCATTTTCACTATGATACTTTCGGGCGCATGCAGGCACTGGAGAATGAACAGGGCGAGCAGTACCGCTTTGAGTATGATGCCTTGCATCGTTTAACCGATGAACATGATCTTATCGGACAACAAAAGCACTATCAGTACGATGTGATGGGGAATGTCACCCAGATAAAAACTACCCCGGGGCCTTGCGTTGATACGCCGATGCCCTTATCCCCGCTGGTGACCACCTTCGGTTACGATAAAGTGGGACGGTTGCTGTTTCGAGAAAACGCCGATTATCGCACGGAATACCTTTATCAACCTTTGAGTGTGACGTTACGCCGAGTCCCCATGGCGATATGGCATGAAGCCGAACGCACCGGCACCACCGCCCGCGTAGAGTATCAAGACGCCCTGACCTTTACCTACGATAAAGTGGGGCAGTTGGTTCGAGAAGCCAGTGCCCGCGATGATTATCAGCATCACTATGATGTGCTGGGTAATATCACTCGCACCGAGTTGCCCCATCAAAGGGCATTTGAATACCTGTATTACGGCTCTGGGCATTTACAACAAACGCAATGGCGAGATAATGCGCAACTGACGGTATTGGCGGAATATCAACGAGACCGTTTACACCGAGAAACGTTGCGCACCAGTGGCGCCTTAGACAATGAAACGGGCTATGACTGTCGGGGACGTATTACGCACCAAGTGGCGCGCCAAATGAATGCCTCACAGTTTGTGACCCCGGTGATTGACCGTCGTTATCGTTGGGATAAACGCAATCAACTTATCGAGCGCAGTGTCAGTTATGGTCAAACAGGCGAGGTTTTTACCGCCGGACATTGGTACTACCACAGTTATCAGTACGACCCGCTGGGGCAATTGACGGCGCATTTAGGGTCGGTGCAAACCGAACACTTTCTGTATGATGCGGCGGCCAATTTACTGACTCGTCCGCACAGCGAAGCACCGCATAACCAAGTACAAGGCAGTGATAAGTATGATTATCGCTACGACGGTTTTGGTCGCATGGTCTCCCGCTATGAAAAAGGCAGTAGCTCGGGACAACGTTATCACTATGACAGTGACCACCGCATTATCGCGGTGGATATCGACCAAGGGCCCTTAGGGTATCAGCGAGCGGAATACCGCTACGATATTTTAGGGCGCCGGATTGAAAAACGGTTATGGAAAGCCAGTGCGATTGCCAACACAGTCACCTATCACCAACATGAGCCGGATGAAGTGTACACTTTTGGCTGGGTAGGGATGCGACTGGTGTCTGAGCACAGCAGCGCAGCGCCCCATACCACGGTTTATCATGCATACAATGACCAAAGTTATACGCCTCTTGCTCGCATTGAATGCACGGATAACCCGCTTAATCCGCAACGGGCGATTTATTATACGCACAGCAGTTTAAGTGGCTTACCGGAAGCGTTGACCAACAGTGAGGGTGAAATAGTTTGGCAAGGGCAATACAGTGTTTGGGGACATTTACAGCGTCAAACCCGCCCCACAAGCACATTTAATCGTGAACAGAACCTGCGCTTTCAAGGGCAGTATTTCGACAAAGAGACGGGGTTACATTATAATACGTTCAGGTACTATGCCCCGGATTTAGGCCGTTTTACCCAGCAAGACCCAATAGGGTTGGCAGGGGGCATAAACTTATATGCTTATGCGCCAAATCCATTGACATGGGTGGATCCGTGGGGTTGGAGTTGCTTTAGTGCTCGGGTAGGTGCTTTTGGTGAGAAAAGAGTTATGAAATACTTATCTGGAGCGGGCTATAAAAAAGTTTTTTCTGTACAAAACAATTCTGGGCATGGTCTGGATATAGTTGCTTTAAGACCAGATGGAAAATTTGATATTTTTGAAGTTAAAAGTTCGACAATAGGACAATTTTCTTTATCTTCCCGCCAAGCTACAGGCGATGATTTTGCAAAAATAGTTCTTTTAAACGATGTGAAAAAAGGAGGTTATAATATTATCGATATAGATGGTAATGTTAAAGCAATTACAAGTAAACAAGCTAGATACATTTATAATAACATAGGAACAACCGAGTGGGTTCAGGTAAATGTTGGCCGAAATCCCAAATTTTATGATCAGAATATAACGTTTGAGGAATGGTAG
<span class="o">&gt;</span>Pseudomonas_fluorescens_F113_idrD
GTGCAGGACGGCAAGAATCGCCTGGTCACCGCCAGCCTCACCAGCCGCGACATCAAGAACAAAGCCGAGTACGACGCCAGCACCGTCAGCGTCGGCGGCGGCTATCGCGAAGTCGGCAAGGACCAGCAGGGCAACGCCACCAGCGGCGGCACGCAAACCCCTGGCACCGACCTGCCGAAGAACGACAACGACCTCAGCGCCACCATGCCGATCGCCATCAGTGCCAGCGACGATGCCAGCAGCGTCACCCGCAGCGGCATCAGTGGCGGCACCATCGTCATCAGCAACGACGCCGAGCAACAGAAGCGCACCGGCAAGAACGGCGAGCAAACTATCGCCAGCCTCAACCGCGATGTCTCCAGCGACCGCGACACCACCAACACGCTCAAGCCGATCTTCGACGAAGACGAAATCCGTGCCGGCTTCGAGATCGTCGGGGCGTTCGCCAACGAAGCCAGCACCTTGCTCGCCAACAAGGCCAAAGAAGTCGACCTCAAGCGCAAACAGGCCGAGGCGGCTCAGACAGCGGCCCTGGACCCCAACGCCAAACTCACCGACACCGAACGCCTGGCGCTGCTCGGCAGCGCTCAGCGCCTGAACGCCGAAGCCAGCGGCATCGCCGACAAGTGGGGCGCGGGCGGTACCTACCGGCAGATCACCACCGCCCTGGTGGGCGCGGCCAGCGGCAACATCAGCGGCGGTAACAGTGCCTTCGTTCAAGGGCTGGTGGTCAATTACGTGCAGCAGCAAGGTGCCGGCTACATCGGCGACCTGGTGGCGGACGGCAAACTCACCGAAGGCTCGCCGCTGCACGCCGCCCTGCACGGCATCGTCGCCTGTGCCGGCGCCGCCGCCAGCAGCCAGAGCTGCGGCAGCGGTGCTGCAGGCGCGGCGGCGTCGAGCCTGCTCACCGGCCTGTTCTCCGAAGCCAGCCCCGATGAAAGCGAATCCCAACGTGAGGCCAAGCGCAACCTGATCGTCAGCATCGTCACCGGGCTGGCCGCCACCAGTTCCTCGCTCGACCCGGCCACCGCCACCAGCGCGGCCGGGGCAGCGGTGGACAACAACTGGCTGGCGACTCAGCAGTTGGTTCAGGCGGAGAAGGAATACGACGCAGCTGACGGCTTAGGGGCCAAGGCGCAAGTGATCGCCAAGTGGGCGTACATCAATCAGAAACAAGACGTCCTGACCCGCTTCGGCGTGGGCAAAGGCTTGGTGCAGGCGGGGTGGAGCGACGTCGAAGGCTTGGCGCAATTCCTCCTTCACCCGGTTGATGGAATCAACGGCCTCAAGGCCCTGATCAACGACCCCGAGGCTCGCAAGCAGTTTGGCGATGGCGTGGTCAACGAACTCAACGCCAAACTTGACCGCGTTAAAGACGCCCTGGAAAACGGTGGTGATGATCGCGCCGTGCAACTGGGCCAAGACATTGGCGAACTCGTCTGGCAGGTCGGCAGCATCGCCACCGGTGTTGGCGGGGCGGCCAAGGGTGGCGTGGCGTTGGCGAAAGCCGGGATCAAGGTGGGGGCGGAAGGGCTTGAAAAAATGGCCGATATGGCCAGGCTTGAAAAAATAGCGAAAGTCGAAGCGCCAGCTACCAAGCCATTGCCTGACTGGAGCGGTGATGGACCCTCCGTCGGGAAGGGGCCAGATAAAACAGTCCCTCCGAGCACGACAACTGATCTGTTTGAAGCGTCTAGAGGTAAGGATCTTTCGACCCTGACAAACCAGCAAATTGGCGACTTGGGTGAGGCGATATCCAAAACCTTCTTGCGCGATAATAACCATACTGACATCTTTGCTATCCAGAACAGTAGCGGAAATGGCATCGATATTGTCAGTCGCACACCTGATGGTCGTCTCGCCTTCACTGAGGTGAAAACATCGCGAACAGGGAGTATTGGAGGGCTTTCCAGTCGTCAAGAAAATATGACGACGTTTGTCGAGGATATTCTGACTCAGGCAGCTAGTCGCCAGGGGCGCTACCGGAATATTGACGTCGCTGATCAACAGCGGGCACGTCAGATGCTCAGGGAGTTTAGGCAAACACCTGAAAGTGTGAGCGGAAATTTGATCGGTGTTGATCTAAAGGGAGAAGTATTGAGAGTCTCTCCTTGGGCGCGCAATTGA
<span class="o">&gt;</span>Rothia_C6B_idrD
atgaccgacatccacggcaatctcttatggtacagcgaatacaccgcctggggtcgtctgaaaaaggacgggcgggtttacaagaacgcgcaccagccgttcagattgcagaaccagtattacgatgaagagaccgggctgcattataacttgttgcgttattacgagcctgaggctgggcggtttgtgaatcaggatccgattgggttgttgggtggggagcatctttatcaatttgcagataacgctttggtttggttcgacccgttgggattgaaaaaaacctacgtgcagcgcttgggtactgccggcgagcgcagggtaatgaaatatctggaaggtacgggatacaaaaaagtcttctctattcaaaatgcctccggaaacggcttggatattgtcgcattgagaccggatgggaaatatgacatatttgaagttaaaagttctaaacgtggaaaatttaaactaagcgaaagacagcagaaaggcggtaaatgttttgctgagcaagtgttgacggaagatgtaacggataagaaaaaaggcggatattttatgaaaggcttagacggcaaaaagacgcctcttaataagaagaaggcgcaagagatatttaataatatagataaaaccgaaactgtctttgttgatatgaatcataaatttcaagcaacccgcatgacatttagcccatggtaa
<span class="o">&gt;</span>Xanthomonas_citri_pv_malvacearum_strain_XcmN1003_idrD
ATGGCCAGCACCACCGAGATCGCAGCACTGCTGTCTGCCACCGATGCACAGGGCCGGGTCGATCCGATGGCCTACGAAGCGCAACGGTTGATCCTGCAGCACCAAGGCCTGGGGAGCGTGGATGCTGCTGCGCTGGTGCATGCTCTGCCGGCCTCGCCGGGATTCGCCACCTTCGACCGTAACGCCTTCTTTTCCGCCATCGACCAGCGCCTGGAGACGCCGCAGGAACGGCAGCGCTTTGCTGATGCGCTGGACCAGGCCAACCTCAGCGATAGCTTGCTCGAGCGGCTGGGCGAGCGGGCCTTTGAGGCCGGTGGGCAGGCCTATGACGCTACGCGCAACGCGGCGAGCCGGGCCGGCGCCCAAGTCAGCGATGGCCTGGCCACTGCGCAGCGCACCGCCCGCCACCCGGCCGCCGATCCCGGTGCCTCGCCCGCGTTGCGCGCTGCAGCGAATGTCGGCAGCAACGTGATCGGTGAGGCTCAGCAAGGCTACGGTTTCGTCACCGGCGGTGCCACGCACGCCCTGTCCACGCTCAGCGATACCGTCGATCTGGCGCGCATGGGTGTGCGCTTGGCGTCCGATCCCAACTACCGAGACGTGGTGGTCGGCATGGTCCGCATCTATGCCGCCGAGGTCGCCGCCGACCCGCACAAGCCGATCGACCAGGTCCGCAGCGCCGTCACGCAGGCGTGGCAGCATTGGGAACAGGGGTTGGCACAGGCGCAACGCGATGGGCGCGAGCGCGAATACATCGGCAGCGCGACCGGTGCGGCCGGAGTGGAGATCCTGGCCACCCTGGTGCCGGCAAGCAAAGTGAGCAAGCTCGGCAAGGCCGCGCAGATCATGGAGGACATCGCCCCCCGGGCCGCCGCCGAGGCCAGCGAAGGCCTGGCCGATGCCGCGCGTGCCGAACGCGCTGGTGCGCGCGCCCACGACGGCACTGAGTATGCGCCCGAGCGCCATGCGGGCTTGCCCGGCCGCGCTCGCGCCTACGCCGAGGGGCTGGAGGAAATCGGCCAGGCCGCCCGGCGTGGTGGTGAGGCCGGCGAAGGGGCGCAGCAGATGCTGCGCGGGCTGGTCGGCCAGGCCCGCCAGGACGGCCACCTGGACGATCTGGTCCGCGCCGCGCGCGCCACCGACAACGTCGAAGGCCTGCTGCGCAGCGGCCAACTGGAACCGCACGAACTCACCGCCATCCTCAAGCGCACGCCGGGTGTCTTTGACGGCAGGATCGATTTCCCGACGGCCCTAGACCACAGTACGCAGGGCGTGGACCTGACCCGGCTCACCACCCGCCAGCTCGGCGACATCGGCGAAGCCATCCAGACCCATGACCTGGTAAAGCAGGGATATAGCGACATCGTCGCGATCAAGAACCGCTCCGGTCACGGCATCGACCTGGTCGGCCGCAACCCCGACGGTGATCTGGAGTTCTTCGAGATTAAGACCTCGGCCAAGGGATTGGCGCCGGCGCAACATGGGGACCCGGAACAATTCGTAGCCAAGCGGCTTGAGCGGGCGATTGACGCCCAGGGTCACTGGGATCCGAAGAACACGATACCTGGGCTGAAGTCGATCGCTCAGGATATCAGCAGGGAAATCGGCCCGGAGGCCGAGAACATCAACGCCACGTGGGTGCGGCTCAATCTGAGCCGGAGCCCGGACGCGCCGCACTTGCAGATCGAACGCAAGATAGATCCATGGGGCAAGCCCGCAGCGCCGAAACAGACGATGCTGACTCCTGGCGATGCCGACCACCCAGACCACGCCGCATTTGACACGATACTGCGCACGGTGCAGGGCGACGGCCGTTGGAACGTTGCGGAAAGCCACAATATTGCCGCGAGCCTGCTGAAGGAATACAAGGCCGATCCGCTGTGTAAGACCCTGGACTCGGTGCGCATCGGCGGCAGCGCCGATACGCCATGCATCTTCGCGGTCAGTTCGCCACACGGCGACAGAGGCCCGGATTTCCATGTCCACGTCCAAGCCGAGCAAGCTGCGCAGCGCCCGGCGGCCGAGAGCTTCGCTCAGGTCCAACAGATCAACCAGCAACAAGCGCTGACCCAGGAAACAGAGCAACAGAACCTGGCTCACCGCGGCCCGAGGATGAGTTGA
</code></pre></div></div>

<p>We put these sequences in a file called <code class="language-plaintext highlighter-rouge">idrD.fna</code>.</p>

<h3 id="identifying-metagenomes-likely-to-contain-populations-with-idrd">Identifying metagenomes likely to contain populations with idrD</h3>

<p>Since there are too many metagenomes to practically download and query
all, we screened for metagenomes likely to contain our sequences (or
similar sequences), using
<a href="https://www.searchsra.org/pages/about">SearchSRA</a>. Briefly, SearchSRA
employs the PARTIE algorithm <a href="https://academic.oup.com/bioinformatics/article/33/15/2389/3092366">(Torres et al
2017)</a>
on subsamples (100,000 reads per metagenome) from all metagenomes on SRA
(which mirrors ENA and other international short read databases) and
queries them with
<a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a>. We
uploaded the FASTA containing all 7 idrD nucleotide sequences to
SearchSRA and searched against all metagenomes.</p>

<p>This resulted in a zipped folder with all the alignments. It was called
<code class="language-plaintext highlighter-rouge">results.zip</code>; we renamed it to <code class="language-plaintext highlighter-rouge">searchsra_results_20190113.zip</code> and
then unzipped it with the command <code class="language-plaintext highlighter-rouge">unzip searchsra_results_20190113.zip</code>
and then entered the resultant directory with
<code class="language-plaintext highlighter-rouge">cd searchsra_results_20190113</code>.</p>

<p>Now, we wanted to find which metagenomes (that is, the *.bam files
produced from PARTIE) had non-zero, or practically zero coverage. Since
PARTIE works by only mapping a [small] subset of reads from each
metagenome, the actual mapped counts are not informative, other than
whether any reads mapped (meaning that metagenome sampled a population
with idrD) or didn’t map. For this first pass filtering of metagenomes,
we worked off of file size, as metagenomes with zero mapping were 4KB,
so we kept anything above 4KB. We did that with this one-liner:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>file <span class="k">in</span> <span class="si">$(</span><span class="nb">ls</span> | <span class="nb">grep</span> <span class="s2">"^[0-9]*$"</span><span class="si">)</span><span class="p">;</span> <span class="k">do 
</span><span class="nb">du</span> <span class="nt">-h</span> <span class="nv">$file</span>/<span class="k">*</span> | <span class="nb">grep</span> <span class="nt">-v</span> <span class="s1">'4.0K'</span><span class="p">;</span> <span class="k">done</span> | <span class="nb">awk</span> <span class="s1">'{print $2}'</span> | <span class="nb">sed</span> <span class="s1">'s/.bam//; s/^[0-9]*\///'</span> 
<span class="o">&gt;</span> sra_run_accessions_gtt4KB.txt
</code></pre></div></div>

<p>This resulted in 3801 metagenome accessions. We downloaded the available
metadata for each of these metagenomes by looping through each entry
like so:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>entry <span class="k">in</span> <span class="si">$(</span><span class="nb">cat </span>sra_run_accessions_gtt4KB.txt<span class="si">)</span><span class="p">;</span> <span class="k">do 

</span>esearch <span class="nt">-db</span> sra <span class="nt">-query</span> <span class="nv">$entry</span> | efetch <span class="nt">-format</span> runinfo <span class="o">&gt;&gt;</span> sra_metadata_accessions_gtt4KB.txt

<span class="k">done</span>
</code></pre></div></div>

<p>This comma-separated file has the needed data but also junk whitespace
(empty lines) so we cleaned that and converted to a tab-separated file
by running</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">grep</span> <span class="nt">-v</span> <span class="s1">'^$'</span> sra_metadata_accessions_gtt4KB.txt | <span class="nb">tr</span> <span class="s1">','</span> <span class="s1">'\t'</span> <span class="o">&gt;</span> temp 
<span class="nb">mv </span>temp sra_metadata_accessions_gtt4KB.txt
</code></pre></div></div>

<p>We took this list and manually curated to remove genome assemblies,
metagenomes with too-specific library preparation methods (e.g. target
capture), etc. to ultimately produce a list of 1,189 metagenomes,
including both single- and paired-end sequence libraries. This table can
be found at <a href="https://drive.google.com/file/d/1iwytSVpvUKMQGmhBuzhby0I8-jT67JFH/view?usp=drive_web">this Drive
link</a>.
In this table, the sample site for each metagenome is listed in the
column “ScientificName”, with varying information content depending on
the uploader.</p>

<h3 id="preparing-the-metagenomes">Preparing the metagenomes</h3>

<h4 id="downloading-and-sorting">Downloading and sorting</h4>

<p>We copied the curated list of 1,189 accessions, just the accessions with
one accession per line, to a text file called
<code class="language-plaintext highlighter-rouge">curated_sra_list_20190114.txt</code> and uploaded it to the cluster. The
first few lines of the file look something like this:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>SRR1182872
SRR948162
SRR1291146
SRR1291147
SRR948162
SRR1291146
SRR1291147
SRR948163
SRR1291148
SRR1291149
</code></pre></div></div>

<p><strong>Note:</strong> Everything from here until the end of <strong>Retrieving the
coverage and variability data</strong>, occured on Harvard’s Odyssey
computational cluster, due to the large filesizes and computational
resources. Odyssey uses the Slurm workload scheduler; batch scripts
copied below retain all header information for maximum reproducibility.</p>

<p>We looped through this list of metagenomes to download each one in
parallel batches, using the following script submitted to Odyssey:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>
<span class="c">#SBATCH -N 1</span>
<span class="c">#SBATCH -n 1</span>
<span class="c">#SBATCH --contiguous</span>
<span class="c">#SBATCH --mem=1G</span>
<span class="c">#SBATCH -t 0-01:00:00</span>
<span class="c">#SBATCH -p shared</span>
<span class="c">#SBATCH --job-name="downloadMaster"</span>
<span class="c">#SBATCH -o odyssey_downloadMaster.out</span>
<span class="c">#SBATCH -e odyssey_downloadMaster.err</span>
<span class="c">#SBATCH --mail-type=NONE</span>

<span class="c"># chunk the accessions into batches</span>
<span class="nb">cat </span>curated_sra_list_20190114.txt | <span class="nb">split</span> <span class="nt">-l</span> 20 <span class="nt">-d</span> - sraBatch_

<span class="c"># get the batches</span>
<span class="nv">batches</span><span class="o">=</span><span class="si">$(</span><span class="nb">ls </span>sraBatch_<span class="k">*</span><span class="si">)</span>

<span class="c"># write and submit the script that does the downloading, for each batch</span>
<span class="k">for </span>batch <span class="k">in</span> <span class="nv">$batches</span><span class="p">;</span> <span class="k">do

</span><span class="nb">echo</span> <span class="s2">"#!/bin/bash
#SBATCH -N 1
#SBATCH -n 1
#SBATCH --contiguous
#SBATCH --mem=1500M
#SBATCH -t 0-06:00:00
#SBATCH -p shared
#SBATCH --job-name='</span><span class="nv">$batch</span><span class="s2">-tar'
#SBATCH -o odyssey_unstuff-</span><span class="nv">$batch</span><span class="s2">.out
#SBATCH -e odyssey_unstuff-</span><span class="nv">$batch</span><span class="s2">.err
#SBATCH --mail-type=NONE

module load sratoolkit

for sra in </span><span class="se">\$</span><span class="s2">(cat </span><span class="nv">$batch</span><span class="s2">); do

# fastq-dump does the downloading, --split-3 dumps 'junk' or unpaired reads a 3rd file
fastq-dump --split-3 </span><span class="nv">$sra</span><span class="s2">

done
"</span> <span class="o">&gt;</span> slurm-<span class="nv">$batch</span>.sh <span class="o">&amp;&amp;</span> sbatch slurm-<span class="nv">$batch</span>.sh

<span class="nb">sleep </span>30

<span class="k">done</span>
</code></pre></div></div>

<p>Some details about this script:</p>

<ul>
  <li>The lines starting with #SBATCH are instructions to the cluster we
used (which uses Slurm)</li>
  <li>This is an ad-hoc parallelization, where the first part does not
actually download but splits the list up into batches of 20
accessions and then generates and submites a script for each batch
to actually download the metagenomes.
    <ul>
      <li>Note: This could be written more efficiently as a job array, but
our cluster was unreliable at this time so retaining individual
batch’s scripts was useful for resubmitting stalled jobs.</li>
    </ul>
  </li>
</ul>

<p>This resulted in 1,188 metagenomes (a single metagenome failed to
download; attempts to download individually failed repeatedly). The next
step was to separate paired-end sequences from single-end sequences as
the mapping program handles them differently. We accomplished this with
a small script:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>

<span class="nb">mkdir </span>single_end
<span class="nb">mkdir </span>paired_end

<span class="c"># move paired ends, clean off 3rd file of errors</span>
<span class="k">for </span>r1 <span class="k">in</span> <span class="si">$(</span><span class="nb">ls</span> <span class="k">*</span>_1.fastq<span class="si">)</span><span class="p">;</span> <span class="k">do</span>
  
  <span class="c"># extract the ID</span>
  <span class="nv">prefix</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="s2">"</span><span class="nv">$r1</span><span class="s2">"</span> | <span class="nb">sed</span> <span class="s1">'s/_1.fastq//'</span><span class="si">)</span>
  
  <span class="c"># move the PE files to paired_end</span>
  <span class="nb">mv</span> <span class="nv">$r1</span> paired_end
  <span class="nb">mv</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>_2.fastq paired_end
  
  <span class="c"># delete any 3rd errors</span>
  <span class="nb">rm</span> <span class="k">${</span><span class="nv">prefix</span><span class="k">}</span>.fastq

  <span class="c"># make list of which samples were pe and where to find them</span>
  <span class="nb">echo</span> <span class="s2">"paired_end/</span><span class="nv">$prefix</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> sra_paired_end.txt  
<span class="k">done</span>

<span class="c"># move single fastqs leftover</span>
<span class="c"># make list of which samples were se and where to find them</span>
<span class="nb">ls</span> <span class="k">*</span>fastq | <span class="nb">sed</span> <span class="s1">'s,.fastq,,; s,^,single_end/,'</span> <span class="o">&gt;&gt;</span> sra_single_end.txt
<span class="nb">mv</span> <span class="k">*</span>fastq single_end
</code></pre></div></div>

<h4 id="mapping">Mapping</h4>

<p>Now that everything was set up, we mapped the metagenomic short reads to
the reference idrD sequences. We chose to use bowtie2, an accurate and
fast gap-sensitive read aligner (Langmead &amp; Salzberg 2012). The first
step is to turn the FASTA file of idrD nucleotide sequences into a
bowtie2-compatible reference database, carried out with the command
<code class="language-plaintext highlighter-rouge">bowtie2-build idrD.fna idrD</code>. Now, we can map each metagenome against
the idrD reference database using the following scripts:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>
<span class="c">#SBATCH -N 1 </span>
<span class="c">#SBATCH -n 16 </span>
<span class="c">#SBATCH --contiguous</span>
<span class="c">#SBATCH --mem=32G</span>
<span class="c">#SBATCH -t 1-12:00:00</span>
<span class="c">#SBATCH -p shared</span>
<span class="c">#SBATCH --array=0-32%33</span>
<span class="c">#SBATCH --job-name="bt-pe"</span>
<span class="c">#SBATCH -o odyssey_map_pe_%a.out</span>
<span class="c">#SBATCH -e odyssey_map_pe_%a.err</span>
<span class="c">#SBATCH --mail-type=END</span>

<span class="c"># split -l 20 -d -a 4 sra_paired_end.txt pe_batch-</span>
<span class="c"># mkdir pe_bt_mapped</span>

<span class="nv">style</span><span class="o">=</span><span class="s2">"pe"</span>  <span class="c"># pe = paired end, se = single end</span>

module load samtools bowtie2 zlib xz
<span class="nv">taskNum</span><span class="o">=</span><span class="si">$(</span><span class="nb">printf</span> %04d <span class="nv">$SLURM_ARRAY_TASK_ID</span><span class="si">)</span>

<span class="c"># get list of sample-runs, chop into per sample batches</span>
<span class="nv">batch</span><span class="o">=</span><span class="k">${</span><span class="nv">style</span><span class="k">}</span>_batch-<span class="nv">$taskNum</span>

<span class="c"># iterate for each participant</span>
<span class="k">for </span>samp <span class="k">in</span> <span class="si">$(</span><span class="nb">cat</span> <span class="nv">$batch</span><span class="si">)</span><span class="p">;</span> <span class="k">do

</span><span class="nv">sampOut</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="s2">"</span><span class="nv">$samp</span><span class="s2">"</span> | <span class="nb">sed</span> <span class="s2">"s,^[a-z_]*,</span><span class="k">${</span><span class="nv">style</span><span class="k">}</span><span class="s2">_bt_mapped,"</span><span class="si">)</span>

<span class="nv">R1</span><span class="o">=</span><span class="k">${</span><span class="nv">samp</span><span class="k">}</span>_1.fastq
<span class="nv">R2</span><span class="o">=</span><span class="k">${</span><span class="nv">samp</span><span class="k">}</span>_2.fastq

<span class="c"># map sample set</span>
bowtie2 <span class="nt">-x</span> idrD <span class="nt">-1</span> <span class="nv">$R1</span> <span class="nt">-2</span> <span class="nv">$R2</span> <span class="nt">--no-unal</span> <span class="nt">-S</span> <span class="nv">$sampOut</span>.sam <span class="nt">--threads</span> 16
samtools view <span class="nt">-S</span> <span class="nt">-b</span> <span class="nv">$sampOut</span>.sam <span class="o">&gt;</span> <span class="nv">$sampOut</span><span class="nt">-RAW</span>.bam
samtools <span class="nb">sort</span> <span class="nv">$sampOut</span><span class="nt">-RAW</span>.bam <span class="o">&gt;</span> <span class="nv">$sampOut</span>.bam
samtools index <span class="nv">$sampOut</span>.bam

<span class="c"># clean as you go</span>
<span class="nb">rm</span> <span class="nv">$sampOut</span>.sam <span class="nv">$sampOut</span><span class="nt">-RAW</span>.bam

<span class="k">done</span>
</code></pre></div></div>

<p>The above script was for the paired-end metagenomes, and the script
below was for the single-end metagenomes:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>
<span class="c">#SBATCH -N 1 </span>
<span class="c">#SBATCH -n 16 </span>
<span class="c">#SBATCH --contiguous</span>
<span class="c">#SBATCH --mem=32G </span>
<span class="c">#SBATCH -t 1-00:00:00</span>
<span class="c">#SBATCH -p shared</span>
<span class="c">#SBATCH --array=0-54%20</span>
<span class="c">#SBATCH --job-name="bt-se"</span>
<span class="c">#SBATCH -o odyssey_map_se_%a.out</span>
<span class="c">#SBATCH -e odyssey_map_se_%a.err</span>
<span class="c">#SBATCH --mail-type=END</span>

<span class="c"># split -l 20 -d -a 4 sra_single_end.txt se_batch-</span>
<span class="c"># mkdir se_bt_mapped</span>

<span class="nv">style</span><span class="o">=</span><span class="s2">"se"</span>  <span class="c"># pe = paired end, se = single end</span>

module load samtools bowtie2 zlib xz
<span class="nv">taskNum</span><span class="o">=</span><span class="si">$(</span><span class="nb">printf</span> %04d <span class="nv">$SLURM_ARRAY_TASK_ID</span><span class="si">)</span>

<span class="c"># get specific batch</span>
<span class="nv">batch</span><span class="o">=</span><span class="k">${</span><span class="nv">style</span><span class="k">}</span>_batch-<span class="nv">$taskNum</span>

<span class="c"># iterate for each sample</span>
<span class="k">for </span>samp <span class="k">in</span> <span class="si">$(</span><span class="nb">cat</span> <span class="nv">$batch</span><span class="si">)</span><span class="p">;</span> <span class="k">do

</span><span class="nv">sampOut</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="s2">"</span><span class="nv">$samp</span><span class="s2">"</span> | <span class="nb">sed</span> <span class="s2">"s,^[a-z_]*,</span><span class="k">${</span><span class="nv">style</span><span class="k">}</span><span class="s2">_bt_mapped,"</span><span class="si">)</span>

<span class="c"># map sample set</span>
bowtie2 <span class="nt">-x</span> idrD <span class="nt">-U</span> <span class="k">${</span><span class="nv">samp</span><span class="k">}</span>.fastq <span class="nt">--no-unal</span> <span class="nt">-S</span> <span class="nv">$sampOut</span>.sam <span class="nt">--threads</span> 16
samtools view <span class="nt">-S</span> <span class="nt">-b</span> <span class="nv">$sampOut</span>.sam <span class="o">&gt;</span> <span class="nv">$sampOut</span><span class="nt">-RAW</span>.bam
samtools <span class="nb">sort</span> <span class="nv">$sampOut</span><span class="nt">-RAW</span>.bam <span class="o">&gt;</span> <span class="nv">$sampOut</span>.bam
samtools index <span class="nv">$sampOut</span>.bam

<span class="c"># clean as you go</span>
<span class="nb">rm</span> <span class="nv">$sampOut</span>.sam <span class="nv">$sampOut</span><span class="nt">-RAW</span>.bam

<span class="k">done</span>
</code></pre></div></div>

<p><strong>IMPORTANT:</strong> Before running these scripts, as they are job arrays, we
ran the first commented-out lines
<code class="language-plaintext highlighter-rouge">split -l 20 -d -a 4 sra_single_end.txt se_batch-</code> and
<code class="language-plaintext highlighter-rouge">mkdir se_bt_mapped</code> for the single-end metagenomes, and the
corresponding lines for the paired-end metagenomes. These commands first
make the lists of files (with 20 accessions per file) that will be
processed by each instance of the job array and secondly make the
directory into which the mapping results will be placed.</p>

<h4 id="importing-the-results-into-anvio">Importing the results into Anvi’o</h4>

<p>We managed the data with anvi’o, a platform for analysis and
visualization of ’omics data <a href="http://merenlab.org/software/anvio/">(Eren et al.,
2015)</a>. The first step is to make a
so-called contigs database to hold the sequence information, which we
accomplished with the command
<code class="language-plaintext highlighter-rouge">anvi-gen-contigs-database -f idrD.fna -o idrD-CONTIGS.db --skip-gene-calling</code></p>

<p>Now we can profile the contigs database with the bowtie2 results, which
will make a so-called profile database that stores and summarizes the
bowtie2 mapping results in a readily accessible manner. For the
paired-end metagenomes, we used this script:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>
<span class="c">#SBATCH -N 1 </span>
<span class="c">#SBATCH -n 4</span>
<span class="c">#SBATCH --contiguous</span>
<span class="c">#SBATCH --mem=2G</span>
<span class="c">#SBATCH -t 0-01:00:00</span>
<span class="c">#SBATCH -p shared</span>
<span class="c">#SBATCH --array=0-32%33</span>
<span class="c">#SBATCH --job-name="prof-pe"</span>
<span class="c">#SBATCH -o odyssey_prof_pe_%a.out</span>
<span class="c">#SBATCH -e odyssey_prof_pe_%a.err</span>
<span class="c">#SBATCH --mail-type=END</span>

<span class="c"># split -l 20 -d -a 4 sra_paired_end.txt pe_batch-</span>
<span class="c"># mkdir pe_profs</span>

<span class="nv">style</span><span class="o">=</span><span class="s2">"pe"</span>  <span class="c"># pe = paired end, se = single end</span>

<span class="c"># activate anvio</span>
<span class="nb">source</span> ~/virtual-envs/anvio-dev-venv/bin/activate
module load samtools bowtie2 zlib xz

<span class="nv">taskNum</span><span class="o">=</span><span class="si">$(</span><span class="nb">printf</span> %04d <span class="nv">$SLURM_ARRAY_TASK_ID</span><span class="si">)</span>

<span class="c"># get list of sample-runs, chop into per sample batches</span>
<span class="nv">batch</span><span class="o">=</span><span class="k">${</span><span class="nv">style</span><span class="k">}</span>_batch-<span class="nv">$taskNum</span>

<span class="c"># iterate for each participant</span>
<span class="k">for </span>samp <span class="k">in</span> <span class="si">$(</span><span class="nb">cat</span> <span class="nv">$batch</span><span class="si">)</span><span class="p">;</span> <span class="k">do

</span><span class="nv">bam</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="s2">"</span><span class="nv">$samp</span><span class="s2">"</span> | <span class="nb">sed</span> <span class="s2">"s,^[a-z_]*,</span><span class="k">${</span><span class="nv">style</span><span class="k">}</span><span class="s2">_bt_mapped,"</span><span class="si">)</span>
<span class="nv">profOut</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="s2">"</span><span class="nv">$samp</span><span class="s2">"</span> | <span class="nb">sed</span> <span class="s2">"s,^[a-z_]*,</span><span class="k">${</span><span class="nv">style</span><span class="k">}</span><span class="s2">_profs,"</span><span class="si">)</span>

<span class="c"># make anvio profiles, -M 0 means don't skip profiling contigs less than 2.5kb</span>
anvi-profile <span class="nt">-i</span> <span class="nv">$bam</span>.bam <span class="nt">-c</span> idrD-CONTIGS.db <span class="nt">-o</span> <span class="nv">$profOut</span> <span class="nt">-M</span> 0 <span class="nt">-T</span> 4 <span class="nt">--write-buffer-size</span> 700

<span class="k">done</span>
</code></pre></div></div>

<p>Note that like before, the <code class="language-plaintext highlighter-rouge">split</code> and <code class="language-plaintext highlighter-rouge">mkdir</code> commands commented out in
the script were run at the command line prior to submitting this array;
likewise for the single-end metagenomes but substituting <code class="language-plaintext highlighter-rouge">single</code> and
<code class="language-plaintext highlighter-rouge">se</code> for <code class="language-plaintext highlighter-rouge">paired</code> and <code class="language-plaintext highlighter-rouge">pe</code>, respectively.</p>

<p>For the single-end metagenomes, we used the same script but changed the
line <code class="language-plaintext highlighter-rouge">style="pe"</code> to <code class="language-plaintext highlighter-rouge">style="se"</code></p>

<p>This resulted in two directories, <code class="language-plaintext highlighter-rouge">pe_profs</code> and <code class="language-plaintext highlighter-rouge">se_profs</code>, each with
hundreds of profiles, one per metagenome, that contained all the
coverage and variability information we are interested in. To facilitate
future steps, we used the <code class="language-plaintext highlighter-rouge">anvi-merge</code> command to merge the indiviual
profile database into a single merged profile database with the
following script:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>
<span class="c">#SBATCH -N 1 </span>
<span class="c">#SBATCH -n 1</span>
<span class="c">#SBATCH --contiguous</span>
<span class="c">#SBATCH --mem=12G</span>
<span class="c">#SBATCH -t 0-04:00:00</span>
<span class="c">#SBATCH -p shared</span>
<span class="c">#SBATCH --job-name="merge"</span>
<span class="c">#SBATCH -o odyssey_merge.out</span>
<span class="c">#SBATCH -e odyssey_merge.err</span>
<span class="c">#SBATCH --mail-type=END</span>

<span class="c"># activate anvio</span>
<span class="nb">source</span> ~/virtual-envs/anvio-dev-venv/bin/activate

<span class="nv">se</span><span class="o">=</span><span class="si">$(</span><span class="nb">cat </span>sra_single_end.txt | <span class="nb">sed</span> <span class="s1">'s/^[a-z_]*/se_profs/'</span><span class="si">)</span>
<span class="nv">pe</span><span class="o">=</span><span class="si">$(</span><span class="nb">cat </span>sra_paired_end.txt | <span class="nb">sed</span> <span class="s1">'s/^[a-z_]*/pe_profs/'</span><span class="si">)</span>

<span class="c"># make anvio profiles, -M 0 means don't skip profiling contigs less than 2.5kb</span>
anvi-merge <span class="nv">$se</span> <span class="nv">$pe</span> <span class="nt">-c</span> idrD-CONTIGS.db <span class="nt">-o</span> idrD-MERGED <span class="nt">-M</span> 0 <span class="nt">--skip-concoct-binning</span>
</code></pre></div></div>

<h4 id="retrieving-the-coverage-and-variability-data">Retrieving the coverage and variability data</h4>

<p>Now that all the data was together, we extracted the coverages and
variability information per site per gene for each metagenome. To do
this, we made a dummy collection to keep anvi’o happy with the command
<code class="language-plaintext highlighter-rouge">anvi-script-add-default-collection -p idrD-MERGED/PROFILE.db</code>. Then, we
exported the coverages with
<code class="language-plaintext highlighter-rouge">anvi-get-split-coverages -c idrD-CONTIGS.db -p idrD-MERGED/PROFILE.db -C DEFAULT -b EVERYTHING -o idrD_covs.tsv</code>
and the variability information with
<code class="language-plaintext highlighter-rouge">anvi-gen-variability-profile -c idrD-CONTIGS.db -p idrD-MERGED/PROFILE.db -C DEFAULT -b EVERYTHING --include-split-names -o idrD_var.tsv</code></p>

<p>Here are first 10 rows of the coverage data:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>nt_position split_name  sample_name coverage
0   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
1   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
2   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
3   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
4   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
5   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
6   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
7   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
8   Acinetobacter_baumannii_strain_XH858_idrD_split_00001   ERR1121387  0
</code></pre></div></div>

<p>And here are the first 10 rows of the variability data:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>entry_id    unique_pos_identifier   pos pos_in_contig   split_name  sample_id   corresponding_gene_call in_partial_gene_call    in_complete_gene_call   base_pos_in_codon   codon_order_in_gene codon_number    gene_length coverage    cov_outlier_in_split    cov_outlier_in_contig   A   C   G   N   T   reference   consensus   competing_nts   departure_from_reference    departure_from_consensus    n2n1ratio   entropy
0   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  DRR046092   -1  0   0   0   -1  0   -1  13  0   0   0   9   4   0   0   C   C   CG  0.3076923076923077  0.3076923076923077  0.4444444444444444  0.6172417697303416
1   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  DRR046097   -1  0   0   0   -1  0   -1  14  0   0   0   10  4   0   0   C   C   CG  0.2857142857142857  0.2857142857142857  0.4 0.5982695885852573
2   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  SRR1284495  -1  0   0   0   -1  0   -1  35  0   0   0   31  0   0   4   C   C   CT  0.11428571428571428 0.11428571428571428 0.12903225806451613 0.355382896246011
3   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  SRR3067121  -1  0   0   0   -1  0   -1  18  0   0   0   14  2   0   2   C   C   CG  0.2222222222222222  0.2222222222222222  0.14285714285714285 0.6837389058487535
4   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  SRR3067123  -1  0   0   0   -1  0   -1  27  0   0   1   21  2   0   3   C   C   CT  0.2222222222222222  0.2222222222222222  0.14285714285714285 0.7544627023259549
5   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  SRR3067131  -1  0   0   0   -1  0   -1  36  1   1   1   32  2   0   1   C   C   CG  0.1111111111111111  0.1111111111111111  0.0625  0.4643566259363562
6   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  SRR3067135  -1  0   0   0   -1  0   -1  54  0   0   0   49  4   0   1   C   C   CG  0.09259259259259259 0.09259259259259259 0.08163265306122448 0.3548290085661212
7   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  SRR3067140  -1  0   0   0   -1  0   -1  13  0   0   1   9   1   0   2   C   C   CT  0.3076923076923077  0.3076923076923077  0.2222222222222222  0.937155853065701
8   0   155 155 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  SRR3067141  -1  0   0   0   -1  0   -1  61  1   1   0   57  1   0   3   C   C   CT  0.06557377049180328 0.06557377049180328 0.05263157894736842 0.27891059970489007
</code></pre></div></div>

<p>The
<a href="https://drive.google.com/file/d/1lMiK2ZuUqwm6sWn6R3UMu5L8jxVgO_hq/view?usp=drive_web">coverage</a>
and
<a href="https://drive.google.com/file/d/1z6RlIJciZZnwmA9arKIuC4A2RVyp6eVZ/view?usp=drive_web">varibility</a>
files were copied down from the cluster and then plotted with R,
described in the next section.</p>

<p>We also ran these analyses for <em>Prevotella</em> spp., described later on in
this document, but the <a href="https://drive.google.com/file/d/1JVuJ5ego8dg_aP--GSca1hnEkC6T3gmA/view?usp=drive_web">coverage data
here</a>
and <a href="https://drive.google.com/file/d/1Ml6ihIBoInAAJz8-QdDCmknBhZMm9eLj/view?usp=drive_web">variability data
here</a>
are here as the code below will process both at the same time.</p>

<h3 id="inspecting-metagenome-coverage">Inspecting metagenome coverage</h3>

<h4 id="preliminary-data-processing">Preliminary data processing</h4>

<p>First, we read in the data like this:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">reshape2</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">plyr</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">entropy</span><span class="p">)</span><span class="w">

</span><span class="n">reads_path</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"~/idrD_analysis/idrD_covs.tsv"</span><span class="w">
</span><span class="n">var_path</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"~/idrD_analysis/idrD_var.tsv"</span><span class="w">
</span><span class="n">metadata_path</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"~/idrD_analysis/sra_metadata_accessions_gtt4KB_mgenomes_abbrev.txt"</span><span class="w">
</span><span class="n">prevo_cov_path</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"~/idrD_analysis/prevo_idrD-covs.tsv"</span><span class="w">  </span><span class="c1"># this is for later</span><span class="w">
</span><span class="n">prevo_var_path</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"~/idrD_analysis/prevo_idrD-var.tsv"</span><span class="w">   </span><span class="c1"># this is for later</span><span class="w">
</span><span class="n">covs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.table</span><span class="p">(</span><span class="n">reads_path</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">vari</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.table</span><span class="p">(</span><span class="n">var_path</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\t"</span><span class="p">)</span><span class="w">

</span><span class="c1"># check out the first few lines</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">covs</span><span class="p">,</span><span class="m">3</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">vari</span><span class="p">,</span><span class="m">3</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##   nt_position                                            split_name sample_name
## 1           0 Acinetobacter_baumannii_strain_XH858_idrD_split_00001  ERR1121387
## 2           1 Acinetobacter_baumannii_strain_XH858_idrD_split_00001  ERR1121387
## 3           2 Acinetobacter_baumannii_strain_XH858_idrD_split_00001  ERR1121387
##   coverage
## 1        0
## 2        0
## 3        0

##   entry_id unique_pos_identifier pos pos_in_contig
## 1        0                     0 155           155
## 2        1                     0 155           155
## 3        2                     0 155           155
##                                                       split_name  sample_id
## 1 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  DRR046092
## 2 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001  DRR046097
## 3 Prevotella_jejuni_strain_CD3_33_chromosome_II_idrD_split_00001 SRR1284495
##   corresponding_gene_call in_partial_gene_call in_complete_gene_call
## 1                      -1                    0                     0
## 2                      -1                    0                     0
## 3                      -1                    0                     0
##   base_pos_in_codon codon_order_in_gene codon_number gene_length coverage
## 1                 0                  -1            0          -1       13
## 2                 0                  -1            0          -1       14
## 3                 0                  -1            0          -1       35
##   cov_outlier_in_split cov_outlier_in_contig A  C G N T reference consensus
## 1                    0                     0 0  9 4 0 0         C         C
## 2                    0                     0 0 10 4 0 0         C         C
## 3                    0                     0 0 31 0 0 4         C         C
##   competing_nts departure_from_reference departure_from_consensus n2n1ratio
## 1            CG                0.3076923                0.3076923 0.4444444
## 2            CG                0.2857143                0.2857143 0.4000000
## 3            CT                0.1142857                0.1142857 0.1290323
##     entropy
## 1 0.6172418
## 2 0.5982696
## 3 0.3553829
</code></pre></div></div>

<p>The file at <code class="language-plaintext highlighter-rouge">metadata_path</code> is the metadata downloaded earlier from NCBI
and is the file we used earlier to get the curated metagenomes list, but
we trimmed off some columns here that we did not use.</p>

<p>Next, we read and processed the metadata to identify which metagenome
came from which category. The metadata called the HMP datasets only as
“human metagenome”, not separating them by body site, but that
information was stored in the “AnalyteType” column. So, we updated the
“ScientificName” column, which held the desired information for all the
other metagenomes, to reflect whether the metagenome was HMP oral or
otherwise. We also manually looked through the “human metagenome”
samples to identify other oral metagenomes that weren’t from the HMP. We
added a column called “category” to the <code class="language-plaintext highlighter-rouge">covs</code> and <code class="language-plaintext highlighter-rouge">vari</code> dataframes to
hold this information.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># drop unique "Run" id's</span><span class="w">
</span><span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">distinct</span><span class="p">(</span><span class="n">read.table</span><span class="p">(</span><span class="n">metadata_path</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\t"</span><span class="p">),</span><span class="w"> </span><span class="n">Run</span><span class="p">,</span><span class="w"> </span><span class="n">.keep_all</span><span class="o">=</span><span class="nb">T</span><span class="p">)</span><span class="w">  
</span><span class="n">rownames</span><span class="p">(</span><span class="n">metadata</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">Run</span><span class="w">
</span><span class="c1"># convert to character to be safe</span><span class="w">
</span><span class="n">metadata</span><span class="o">$</span><span class="n">ScientificName</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">ScientificName</span><span class="p">)</span><span class="w">  
</span><span class="c1"># rename the HMP gut samples' ScientificName</span><span class="w">
</span><span class="n">metadata</span><span class="o">$</span><span class="n">ScientificName</span><span class="p">[</span><span class="nf">as.character</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">Analyte_Type</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w"> 
                          </span><span class="s2">"G_DNA_STOOL"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"HMP gut metagenome"</span><span class="w">  
</span><span class="c1"># likewise for HMP oral</span><span class="w">
</span><span class="n">metadata</span><span class="o">$</span><span class="n">ScientificName</span><span class="p">[</span><span class="n">metadata</span><span class="o">$</span><span class="n">Analyte_Type</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                          </span><span class="nf">c</span><span class="p">(</span><span class="s2">"G_DNA_Tongue dorsum"</span><span class="p">,</span><span class="s2">"G_DNA_Throat"</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"HMP oral metagenome"</span><span class="w">  
</span><span class="n">metadata</span><span class="o">$</span><span class="n">ScientificName</span><span class="p">[</span><span class="n">metadata</span><span class="o">$</span><span class="n">Run</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                          </span><span class="nf">c</span><span class="p">(</span><span class="s2">"SRR5892193"</span><span class="p">,</span><span class="s2">"SRR5892208"</span><span class="p">,</span><span class="s2">"SRR5892209"</span><span class="p">,</span><span class="w">
                            </span><span class="s2">"SRR5892216"</span><span class="p">,</span><span class="s2">"SRR5892235"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SRR5892239"</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Non-HMP oral metagenome"</span><span class="w">
</span><span class="n">metadata</span><span class="o">$</span><span class="n">ScientificName</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">ScientificName</span><span class="p">)</span><span class="w">  </span><span class="c1"># convert back from character to factor</span><span class="w">

</span><span class="c1"># add in metadata, gene type</span><span class="w">
</span><span class="n">covs</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="p">[</span><span class="nf">as.character</span><span class="p">(</span><span class="n">covs</span><span class="o">$</span><span class="n">sample_name</span><span class="p">),</span><span class="s2">"ScientificName"</span><span class="p">]</span><span class="w">
</span><span class="n">vari</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="p">[</span><span class="nf">as.character</span><span class="p">(</span><span class="n">vari</span><span class="o">$</span><span class="n">sample_id</span><span class="p">),</span><span class="s2">"ScientificName"</span><span class="p">]</span><span class="w">
</span><span class="n">covs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unite</span><span class="p">(</span><span class="n">covs</span><span class="p">,</span><span class="w"> </span><span class="s2">"sample_split"</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"sample_name"</span><span class="p">,</span><span class="s2">"split_name"</span><span class="p">),</span><span class="w"> </span><span class="n">remove</span><span class="o">=</span><span class="nb">F</span><span class="p">)</span><span class="w">

</span><span class="c1"># do the same for prevo</span><span class="w">
</span><span class="n">prevo_cov</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.table</span><span class="p">(</span><span class="n">prevo_cov_path</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">prevo_var</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.table</span><span class="p">(</span><span class="n">prevo_var_path</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">prevo_cov</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="p">[</span><span class="nf">as.character</span><span class="p">(</span><span class="n">prevo_cov</span><span class="o">$</span><span class="n">sample_name</span><span class="p">),</span><span class="s2">"ScientificName"</span><span class="p">]</span><span class="w">
</span><span class="n">prevo_var</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="p">[</span><span class="nf">as.character</span><span class="p">(</span><span class="n">prevo_var</span><span class="o">$</span><span class="n">sample_id</span><span class="p">),</span><span class="s2">"ScientificName"</span><span class="p">]</span><span class="w">
</span><span class="n">prevo_cov</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unite</span><span class="p">(</span><span class="n">prevo_cov</span><span class="p">,</span><span class="w"> </span><span class="s2">"sample_split"</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"sample_name"</span><span class="p">,</span><span class="s2">"split_name"</span><span class="p">),</span><span class="w"> </span><span class="n">remove</span><span class="o">=</span><span class="nb">F</span><span class="p">)</span><span class="w">

</span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">covs</span><span class="o">$</span><span class="n">sample_name</span><span class="p">)))</span><span class="w">  </span><span class="c1"># this many metagenomes</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] 1188
</code></pre></div></div>

<p>Over the course of our investigations (i.e., after some preliminary
plots), we noticed some outlier samples that we removed for various
reasons detailed in the manuscrtipt, and so we drop them here:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">prevo_cov</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">prevo_cov</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">prevo_cov</span><span class="o">$</span><span class="n">sample_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR628272"</span><span class="p">),]</span><span class="w"> </span><span class="c1"># prevo, low quality</span><span class="w">
</span><span class="n">prevo_var</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">prevo_var</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">prevo_var</span><span class="o">$</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR628272"</span><span class="p">),]</span><span class="w">
</span><span class="n">covs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">covs</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">covs</span><span class="o">$</span><span class="n">sample_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR628272"</span><span class="p">),]</span><span class="w"> </span><span class="c1"># prevotella, low quality</span><span class="w">
</span><span class="n">vari</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vari</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">vari</span><span class="o">$</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR628272"</span><span class="p">),]</span><span class="w">
</span><span class="n">covs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">covs</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">covs</span><span class="o">$</span><span class="n">sample_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR1779144"</span><span class="p">),]</span><span class="w"> </span><span class="c1"># cronobacter, infant, skews y-axis</span><span class="w">
</span><span class="n">vari</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vari</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">vari</span><span class="o">$</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR1779144"</span><span class="p">),]</span><span class="w">
</span><span class="n">covs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">covs</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">covs</span><span class="o">$</span><span class="n">sample_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR2047620"</span><span class="p">),]</span><span class="w"> </span><span class="c1"># cronobacter, stem cell kids, skews y-axis</span><span class="w">
</span><span class="n">vari</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vari</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">vari</span><span class="o">$</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR2047620"</span><span class="p">),]</span><span class="w">
</span><span class="n">covs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">covs</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">covs</span><span class="o">$</span><span class="n">sample_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR1038387"</span><span class="p">),]</span><span class="w"> </span><span class="c1"># proteus, neonatal NEC, skews y-axis</span><span class="w">
</span><span class="n">vari</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vari</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">vari</span><span class="o">$</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR1038387"</span><span class="p">),]</span><span class="w">

</span><span class="n">covs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">covs</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">covs</span><span class="o">$</span><span class="n">sample_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR1781983"</span><span class="p">),]</span><span class="w"> </span><span class="c1"># drop the high rothia sample that looks ok but skews y axis</span><span class="w">
</span><span class="n">vari</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vari</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">vari</span><span class="o">$</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR1781983"</span><span class="p">),]</span><span class="w">
</span></code></pre></div></div>

<p>At this point, we have <strong>1183</strong> metagenomes from <strong>45</strong> categories:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] "Number of metagenomes: 1183"

## [1] "Number of sampled categories: 45"

##  [1] "Homo sapiens"                       "metagenomes"                       
##  [3] "metagenome"                         "human gut metagenome"              
##  [5] "gut metagenome"                     "sediment metagenome"               
##  [7] "Homo sapiens neanderthalensis"      "marine metagenome"                 
##  [9] "root"                               "HMP oral metagenome"               
## [11] "human metagenome"                   "human skin metagenome"             
## [13] "hydrothermal vent metagenome"       "terrestrial metagenome"            
## [15] "oral metagenome"                    "bioreactor metagenome"             
## [17] "food metagenome"                    "human oral metagenome"             
## [19] "indoor metagenome"                  "Non-HMP oral metagenome"           
## [21] "insect metagenome"                  "wastewater metagenome"             
## [23] "soil metagenome"                    "oyster metagenome"                 
## [25] "fossil metagenome"                  "pig gut metagenome"                
## [27] "Gallus gallus"                      "bovine gut metagenome"             
## [29] "Bos indicus"                        "freshwater metagenome"             
## [31] "anaerobic digester metagenome"      "hypersaline lake metagenome"       
## [33] "groundwater metagenome"             "microbial mat metagenome"          
## [35] "milk metagenome"                    "upper respiratory tract metagenome"
## [37] "biogas fermenter metagenome"        "compost metagenome"                
## [39] "hot springs metagenome"             "bovine metagenome"                 
## [41] "marine sediment metagenome"         "soda lake metagenome"              
## [43] "surface metagenome"                 "aquatic metagenome"                
## [45] "invertebrate gut metagenome"
</code></pre></div></div>

<h4 id="filtering-metagenomes-based-on-category">Filtering metagenomes based on category</h4>

<p>The 1183 metagenomes retained still need to be filtered, as we wanted to
minimize reporting spurious coverage from metagenomes not representative
of that idrD sequence. In other words, we wanted to investigate only
metagenomes that seem to represent the idrD sequence of interest - while
one sample may contain, say, <em>Prevotella</em> spp. population(s) with idrD
gene, that sample might not contain any <em>Cronobacter turicensis</em> cells,
so including that sample for every sequence would just add noise.</p>

<p>So, we employed a filtering criterion that, for each sequence, discarded
metagenomes for which more than half of the nucleotide positions in that
idrD sequence were <strong>not</strong> covered at least once. Note that this is
applied to each sequence separately; thus, the metagenomes retained for
each sequence are not tied to each other. Here is the function we wrote
to do that:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">filter_mostly_zeros</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="o">=</span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dcast</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">sample_split</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">nt_position</span><span class="p">,</span><span class="w"> </span><span class="n">value.var</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'coverage'</span><span class="p">)</span><span class="w">
    </span><span class="n">rownames</span><span class="p">(</span><span class="n">df_wide</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">df_wide</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="w">

    </span><span class="c1"># the !is.na part  as differing gene lengths between taxa so the dcast padded lengths with na to make square</span><span class="w">
    </span><span class="c1"># count positions covered at least once</span><span class="w">
    </span><span class="n">proportions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">df_wide</span><span class="p">[,</span><span class="m">2</span><span class="o">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">df_wide</span><span class="p">)],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> 
                         </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="m">1</span><span class="o">*</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)]</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="o">/</span><span class="nf">length</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)]))</span><span class="w">  
    </span><span class="n">keepThese</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">proportions</span><span class="p">)[</span><span class="n">proportions</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="p">]</span><span class="w">  </span><span class="c1"># keep samples with enough bases covered</span><span class="w">
    </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">df</span><span class="p">[</span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">sample_split</span><span class="p">)</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">keepThese</span><span class="p">,]</span><span class="w">  </span><span class="c1"># subset</span><span class="w">
    </span><span class="nf">return</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre></div></div>

<p>We also wrote a small function to bin categories together. This will be
relevant later. This function takes a dataframe (like <code class="language-plaintext highlighter-rouge">covs</code>) and a
named list, where each list element’s name is the desired name of the
bin to be made (e.g. <code class="language-plaintext highlighter-rouge">'Human oral metagenomes'</code>) and that list element
is a vector of the categories to be combined into that bin (e.g.,
<code class="language-plaintext highlighter-rouge">c("HMP oral metagenome", "Non-HMP oral metagenome"))</code>).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">bin_categories</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="p">){</span><span class="w">
    </span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="p">)</span><span class="w">
    </span><span class="k">for</span><span class="p">(</span><span class="n">b</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">bins</span><span class="p">)){</span><span class="w">
        </span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="p">[</span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">bins</span><span class="p">[[</span><span class="n">b</span><span class="p">]]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">b</span><span class="w">
    </span><span class="p">}</span><span class="w">
    </span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="p">)))</span><span class="w">
    </span><span class="nf">return</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre></div></div>

<p>With these, we generated a summary heatmap of all of the data. This is
the function to generate the figure:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot_summary_heatmap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="o">=</span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">

    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">bins</span><span class="p">)){</span><span class="w">
        </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bin_categories</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">
    
    </span><span class="c1"># filter non-relevant metagenomes</span><span class="w">
    </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filter_mostly_zeros</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="p">)</span><span class="w">
    
    </span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_idrD.*$"</span><span class="p">,</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">split_name</span><span class="p">)))</span><span class="w">
    </span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">)))</span><span class="w">

    </span><span class="c1"># summarize to get log10 mean coverage by category</span><span class="w">
    </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ddply</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="o">~</span><span class="n">tax</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">category</span><span class="p">,</span><span class="w"> </span><span class="n">summarise</span><span class="p">,</span><span class="w"> </span><span class="n">LogMeanCov</span><span class="o">=</span><span class="n">log10</span><span class="p">(</span><span class="n">mean</span><span class="p">(</span><span class="n">coverage</span><span class="p">)))</span><span class="w">

    </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">category</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">geom_tile</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">LogMeanCov</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">scale_fill_gradient</span><span class="p">(</span><span class="n">low</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'black'</span><span class="p">,</span><span class="w"> </span><span class="n">high</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'#00FF00'</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">xlab</span><span class="p">(</span><span class="s2">"idrD sequences"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ylab</span><span class="p">(</span><span class="s2">"Metagenome categories"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">labs</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">"Log10 Mean Coverage"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">scale_x_discrete</span><span class="p">(</span><span class="n">expand</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">scale_y_discrete</span><span class="p">(</span><span class="n">expand</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">theme</span><span class="p">(</span><span class="n">panel.grid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w"> </span><span class="n">axis.line</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'black'</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w"> 
              </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">colour</span><span class="o">=</span><span class="s1">'black'</span><span class="p">),</span><span class="w"> </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w"> 
              </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="o">=</span><span class="s1">'black'</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">0.3</span><span class="p">),</span><span class="w">
              </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'grey50'</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'black'</span><span class="p">))</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre></div></div>

<p>and this is how we called it to make the plot:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot_summary_heatmap</span><span class="p">(</span><span class="n">covs</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-23-1.png" alt="" /></p>

<p>This heatmap shows, for each gene, the mean coverage of that gene for
metagenomes from each category. Rows are metagenome categories, columns
are idrD sequences from the different species, and each cell is shaded
by the mean coverage over that entire gene from all metagenomes in that
category, log10 transformed. From this plot, it becomes apparent that
the various human microbiomes are interesting, and that idrD sequences
from the <em>Acinetobacter</em>, <em>Pseudomonas</em>, and <em>Xanthomonas</em> spp. are not
particularly well covered in any metagenome, except in metagenomes with
evenly high coverage across all sequences (e.g. biogas fermenter
metagenomes). These species’ sequences will ultimately be dropped, to
focus on <em>Proteus</em>, <em>Rothia</em>, <em>Cronobacter</em>, and <em>Prevotella</em>.</p>

<h4 id="binning-metagenomes-by-origin">Binning metagenomes by origin</h4>

<p>Based on the heatmap, we decided to focus on human oral and human gut
metagenomes. We investigated a few different binning strageties, the one
ultimately used was <code class="language-plaintext highlighter-rouge">bin</code>, the other two represent bins that distinguish
HMP oral from other oral metganomes (<code class="language-plaintext highlighter-rouge">binSplitOralHMP</code>) and include the
FijiCOMP dataset (<code class="language-plaintext highlighter-rouge">binWithFiji</code>).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># metagenome bin categories</span><span class="w">
</span><span class="n">bin</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span><span class="w">
</span><span class="n">bin</span><span class="p">[[</span><span class="s2">"aquatic"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"soda lake metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"marine sediment metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"marine metagenome"</span><span class="p">,</span><span class="w"> 
                      </span><span class="s2">"hypersaline lake metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"hydrothermal vent metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"hot springs metagenome"</span><span class="p">,</span><span class="w"> 
                      </span><span class="s2">"groundwater metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"freshwater metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"aquatic metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">bin</span><span class="p">[[</span><span class="s2">"soil_probably"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"terrestrial metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"surface metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"soil metagenome"</span><span class="p">,</span><span class="w"> 
                            </span><span class="s2">"sediment metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">bin</span><span class="p">[[</span><span class="s2">"animal"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"pig gut metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"oyster metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"invertebrate gut metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Gallus gallus"</span><span class="p">,</span><span class="w"> 
                     </span><span class="s2">"bovine metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"bovine gut metagenome"</span><span class="p">,</span><span class="w">
                     </span><span class="s2">"Bos indicus"</span><span class="p">,</span><span class="w"> </span><span class="s2">"gut metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">bin</span><span class="p">[[</span><span class="s2">"other"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"wastewater metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"upper respiratory tract metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"milk metagenome"</span><span class="p">,</span><span class="w"> 
                    </span><span class="s2">"microbial mat metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"metagenome"</span><span class="p">,</span><span class="s2">"metagenomes"</span><span class="p">,</span><span class="w"> </span><span class="s2">"indoor metagenome"</span><span class="p">,</span><span class="w"> 
                    </span><span class="s2">"Homo sapiens"</span><span class="p">,</span><span class="w"> </span><span class="s2">"fossil metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"food metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"compost metagenome"</span><span class="p">,</span><span class="w"> 
                    </span><span class="s2">"bioreactor metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"biogas fermenter metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"anaerobic digester metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">bin</span><span class="p">[[</span><span class="s2">"Non-HMP gut metagenome"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"human gut metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"human metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">binSplitOralHMP</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bin</span><span class="w">
</span><span class="n">bin</span><span class="p">[[</span><span class="s2">"Human oral metagenome"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"human oral metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"HMP oral metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Non-HMP oral metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">binWithFiji</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bin</span><span class="w">
</span><span class="n">binWithFiji</span><span class="p">[[</span><span class="s2">"Human oral metagenome"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">binWithFiji</span><span class="p">[[</span><span class="s2">"Human oral metagenome"</span><span class="p">]],</span><span class="w"> </span><span class="s2">"oral metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">binSplitOralHMP</span><span class="p">[[</span><span class="s2">"Non-HMP oral metagenome"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Non-HMP oral metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"human oral metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">binSplitOralHMP</span><span class="p">[[</span><span class="s2">"HMP oral metagenome"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"HMP oral metagenome"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>When we bin, the heatmap now looks like this, which is Supplemental
Figure 3 of the paper:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot_summary_heatmap</span><span class="p">(</span><span class="n">covs</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bin</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-25-1.png" alt="" /></p>

<h4 id="per-nucleotide-coverage-to-visualize-region-specific-coverage">Per-nucleotide coverage to visualize region-specific coverage</h4>

<p>Because the heatmap summarizes all nucleotide positions across many
metagenomes, we wanted to understand idrD representation in metagenome
populations at a finer level of detail. This also provides an
opportunity to gauge idrD variability in complicated populations, which
may have implications for its role in subpopulation differentiation, as
suspected from its role in interstrain competition.</p>

<p>So, we wrote a function to plot individual metagenome coverages for each
gene for each habitat we were interested in. Here is the function:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot_mean_covs_var_combo</span><span class="w"> </span><span class="o">&lt;-</span><span class="k">function</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">v</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">subTaxPos</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="w"> 
                                    </span><span class="n">subsite</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">min_var_cov</span><span class="o">=</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">scale_axes</span><span class="o">=</span><span class="s1">'free'</span><span class="p">,</span><span class="w"> </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                                    </span><span class="n">rhsBox</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="n">rhsCols</span><span class="o">=</span><span class="s2">"darkgreen"</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    
    </span><span class="c1"># subset to certain metagenome categories if requested</span><span class="w">
    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">subsite</span><span class="p">)){</span><span class="w">
        </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">subsite</span><span class="p">,</span><span class="w"> </span><span class="n">unlist</span><span class="p">(</span><span class="n">bins</span><span class="p">[</span><span class="n">subsite</span><span class="p">])),]</span><span class="w">
        </span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">v</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">subsite</span><span class="p">,</span><span class="w"> </span><span class="n">unlist</span><span class="p">(</span><span class="n">bins</span><span class="p">[</span><span class="n">subsite</span><span class="p">])),]</span><span class="w">
    </span><span class="p">}</span><span class="w">

    </span><span class="c1"># make taxonomy column with pretty names</span><span class="w">
    </span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_idrD.*$"</span><span class="p">,</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">split_name</span><span class="p">)))</span><span class="w">
    </span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sort</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">))))</span><span class="w">

    </span><span class="n">v</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_idrD.*$"</span><span class="p">,</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">split_name</span><span class="p">)))</span><span class="w">
    </span><span class="n">v</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sort</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">tax</span><span class="p">))))</span><span class="w">

    </span><span class="c1"># subset to certain nuclotide windows</span><span class="w">
    </span><span class="n">print</span><span class="p">(</span><span class="s2">"subsetting..."</span><span class="p">)</span><span class="w">
    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">subTaxPos</span><span class="p">)){</span><span class="w">
        </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">df</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">subTaxPos</span><span class="p">),</span><span class="n">collapse</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"|"</span><span class="p">),</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">split_name</span><span class="p">)),]</span><span class="w">
        </span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">v</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">subTaxPos</span><span class="p">),</span><span class="n">collapse</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"|"</span><span class="p">),</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">split_name</span><span class="p">)),]</span><span class="w">
        </span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">subg</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">subTaxPos</span><span class="p">)){</span><span class="w">
            </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="nf">is.numeric</span><span class="p">(</span><span class="n">subTaxPos</span><span class="p">[[</span><span class="n">subg</span><span class="p">]]))</span><span class="w"> </span><span class="p">{</span><span class="w">
                </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filter</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="o">!</span><span class="p">(</span><span class="n">grepl</span><span class="p">(</span><span class="n">subg</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">split_name</span><span class="p">))</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="n">nt_position</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">subTaxPos</span><span class="p">[[</span><span class="n">subg</span><span class="p">]])))</span><span class="w">
                </span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filter</span><span class="p">(</span><span class="n">v</span><span class="p">,</span><span class="w"> </span><span class="o">!</span><span class="p">(</span><span class="n">grepl</span><span class="p">(</span><span class="n">subg</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">split_name</span><span class="p">))</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="n">pos</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">subTaxPos</span><span class="p">[[</span><span class="n">subg</span><span class="p">]])))</span><span class="w">
            </span><span class="p">}</span><span class="w">
        </span><span class="p">}</span><span class="w">
        </span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sapply</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">subTaxPos</span><span class="p">),</span><span class="w"> </span><span class="c1"># reorder by custom value</span><span class="w">
                                                 </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">),</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)))</span><span class="w">  
        </span><span class="n">v</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">levels</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">))</span><span class="w">  </span><span class="c1"># reorder to match</span><span class="w">
    </span><span class="p">}</span><span class="w">
    </span><span class="n">print</span><span class="p">(</span><span class="s2">"done subsetting"</span><span class="p">)</span><span class="w">

    </span><span class="n">print</span><span class="p">(</span><span class="s2">"about to filter"</span><span class="p">)</span><span class="w">
    </span><span class="k">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">proportion_gte0</span><span class="p">)){</span><span class="w">
        </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filter_mostly_zeros</span><span class="p">(</span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">
    </span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">v</span><span class="p">[</span><span class="nf">as.character</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">sample_id</span><span class="p">)</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">sample_name</span><span class="p">)),]</span><span class="w">  </span><span class="c1"># keep these the same</span><span class="w">
    </span><span class="n">print</span><span class="p">(</span><span class="s1">'done filtering'</span><span class="p">)</span><span class="w">

    </span><span class="c1"># bin metagenome categories if needed</span><span class="w">
    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">bins</span><span class="p">)){</span><span class="w">
        </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bin_categories</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="p">)</span><span class="w">
        </span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bin_categories</span><span class="p">(</span><span class="n">v</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">

    </span><span class="c1"># count number of metagenomes to display for each sequence x metagenome combination</span><span class="w">
    </span><span class="n">numMgenomes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ddply</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="o">~</span><span class="n">tax</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">category</span><span class="p">,</span><span class="w"> </span><span class="n">summarise</span><span class="p">,</span><span class="w"> </span><span class="n">nMeta</span><span class="o">=</span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">sample_name</span><span class="p">))))</span><span class="w">

    </span><span class="c1"># order the metagenome categories as specified</span><span class="w">
    </span><span class="k">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">subsite</span><span class="p">)){</span><span class="w">
        </span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">category</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subsite</span><span class="p">)</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">category</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subsite</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">

    </span><span class="c1"># drop any variant positions if coverage is too low (default = don't drop any variant positions)</span><span class="w">
    </span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">v</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">min_var_cov</span><span class="p">,]</span><span class="w">
    
    </span><span class="c1">## re-compute the Shannon Entropy of SNPs from each position</span><span class="w">
    </span><span class="c1"># requires adding in counts of non-variant metagenomes and then re-calculating entropy</span><span class="w">
    </span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="n">v</span><span class="p">,</span><span class="w"> </span><span class="n">by.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'split_name'</span><span class="p">,</span><span class="w"> </span><span class="s1">'sample_name'</span><span class="p">,</span><span class="w"> </span><span class="s1">'nt_position'</span><span class="p">),</span><span class="w"> 
               </span><span class="n">by.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"split_name"</span><span class="p">,</span><span class="s2">"sample_id"</span><span class="p">,</span><span class="s2">"pos"</span><span class="p">),</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="w">
    </span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">v</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="p">),]</span><span class="w">
    </span><span class="n">v</span><span class="o">$</span><span class="n">category</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">category.x</span><span class="w">

    </span><span class="c1"># for positions where only some metagenomes had variance, add back the difference</span><span class="w">
    </span><span class="n">v</span><span class="o">$</span><span class="n">A</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"A"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> 
      </span><span class="n">v</span><span class="o">$</span><span class="n">A</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"A"</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"A"</span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"A"</span><span class="p">]</span><span class="w">
    </span><span class="n">v</span><span class="o">$</span><span class="n">G</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"G"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> 
      </span><span class="n">v</span><span class="o">$</span><span class="n">G</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"G"</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"G"</span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"G"</span><span class="p">]</span><span class="w">
    </span><span class="n">v</span><span class="o">$</span><span class="nb">T</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"T"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">
      </span><span class="n">v</span><span class="o">$</span><span class="nb">T</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"T"</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"T"</span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"T"</span><span class="p">]</span><span class="w">
    </span><span class="n">v</span><span class="o">$</span><span class="n">C</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"C"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> 
      </span><span class="n">v</span><span class="o">$</span><span class="n">C</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"C"</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"C"</span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w">
        </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="p">[</span><span class="n">v</span><span class="o">$</span><span class="n">coverage.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">coverage.y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">v</span><span class="o">$</span><span class="n">reference</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"C"</span><span class="p">]</span><span class="w">

    </span><span class="c1"># calculate the entropy for each site</span><span class="w">
    </span><span class="n">variability</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ddply</span><span class="p">(</span><span class="n">v</span><span class="p">,</span><span class="w"> </span><span class="o">~</span><span class="n">nt_position</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">tax.x</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">category</span><span class="p">,</span><span class="w"> </span><span class="n">summarise</span><span class="p">,</span><span class="w"> </span><span class="n">origEntropy</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">entropy</span><span class="p">),</span><span class="w"> 
                         </span><span class="n">ent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">entropy.empirical</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">A</span><span class="p">),</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">C</span><span class="p">),</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">G</span><span class="p">),</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nb">T</span><span class="p">)),</span><span class="w"> </span><span class="n">unit</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'log2'</span><span class="p">))</span><span class="w">

    </span><span class="c1"># sort the taxa in the specified order</span><span class="w">
    </span><span class="n">variability</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">variability</span><span class="o">$</span><span class="n">tax.x</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">levels</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">))</span><span class="w">


    </span><span class="c1"># coverage plot</span><span class="w">
    </span><span class="n">q</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nt_position</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">coverage</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">geom_line</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.4</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sample_name</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">tax</span><span class="p">),</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
        </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">numMgenomes</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="n">nMeta</span><span class="p">),</span><span class="w"> </span><span class="n">x</span><span class="o">=</span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">category</span><span class="o">~</span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scale_axes</span><span class="p">,</span><span class="w"> </span><span class="n">switch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"y"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
        </span><span class="n">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">guides</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">xlab</span><span class="p">(</span><span class="s2">"Nucleotide position in idrD"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ylab</span><span class="p">(</span><span class="s2">"Metagenome coverage (X)"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">scale_x_continuous</span><span class="p">(</span><span class="n">expand</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">expand</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">theme</span><span class="p">(</span><span class="n">panel.grid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w"> </span><span class="n">axis.line</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'black'</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w"> 
              </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">colour</span><span class="o">=</span><span class="s1">'black'</span><span class="p">),</span><span class="w"> </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="o">=</span><span class="s1">'black'</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">0.3</span><span class="p">),</span><span class="w"> 
              </span><span class="n">strip.background.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'#dcdcdc'</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">),</span><span class="w">
              </span><span class="n">strip.background.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'#ffffff'</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="o">=</span><span class="kc">NA</span><span class="p">),</span><span class="w"> 
              </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">'black'</span><span class="p">),</span><span class="w"> </span><span class="n">strip.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'bold'</span><span class="p">))</span><span class="w">
    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cols</span><span class="p">)){</span><span class="w">
        </span><span class="n">q</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">q</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cols</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">

    </span><span class="c1"># var plot</span><span class="w">
    </span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">variability</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nt_position</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ent</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">geom_col</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">df</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nt_position</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">coverage</span><span class="o">*</span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">geom_col</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'grey20'</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">category</span><span class="o">~</span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scale_axes</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">scale_x_continuous</span><span class="p">(</span><span class="n">expand</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">expand</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'right'</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">xlab</span><span class="p">(</span><span class="s2">"Nucleotide position in idrD"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ylab</span><span class="p">(</span><span class="s2">"Shannon Entropy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
        </span><span class="n">theme</span><span class="p">(</span><span class="n">panel.grid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w"> </span><span class="n">axis.line</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'black'</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w"> 
              </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">colour</span><span class="o">=</span><span class="s1">'black'</span><span class="p">),</span><span class="w"> </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="o">=</span><span class="s1">'black'</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">0.3</span><span class="p">),</span><span class="w"> 
              </span><span class="n">strip.background.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'#dcdcdc'</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">),</span><span class="w">
              </span><span class="n">strip.background.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'#ffffff'</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="o">=</span><span class="kc">NA</span><span class="p">),</span><span class="w"> 
              </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">'black'</span><span class="p">),</span><span class="w"> </span><span class="n">strip.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'bold'</span><span class="p">))</span><span class="w">
    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">rhsBox</span><span class="p">)){</span><span class="w">
        </span><span class="n">rhsBox</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">tax</span><span class="o">=</span><span class="nf">names</span><span class="p">(</span><span class="n">rhsBox</span><span class="p">),</span><span class="n">matrix</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">rhsBox</span><span class="p">),</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">rhsBox</span><span class="p">),</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">byrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">))</span><span class="w">
        </span><span class="n">rhsBox</span><span class="o">$</span><span class="n">tax_distinct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rhsBox</span><span class="o">$</span><span class="n">tax</span><span class="w">
        </span><span class="n">rhsBox</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sapply</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">rhsBox</span><span class="o">$</span><span class="n">tax</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">),</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">))</span><span class="w">
        </span><span class="n">rhsBox</span><span class="o">$</span><span class="n">tax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">rhsBox</span><span class="o">$</span><span class="n">tax</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">levels</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">tax</span><span class="p">))</span><span class="w">
        </span><span class="n">q</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">q</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_segment</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">rhsBox</span><span class="p">,</span><span class="w"> 
                              </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">X1</span><span class="p">,</span><span class="w"> </span><span class="n">xend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">X2</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">yend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">Inf</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rhsCols</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">


    </span><span class="n">print</span><span class="p">(</span><span class="n">q</span><span class="p">)</span><span class="w">
    </span><span class="n">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre></div></div>

<p>We manually defined a few additional variables to specify the taxa,
sites, and regions of interest. Here they are:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># C terminal domains from taxa we care about, see ms for how start/stop positions were identified</span><span class="w">
</span><span class="n">sub4cterm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span><span class="w">
</span><span class="n">sub4cterm</span><span class="p">[[</span><span class="s2">"Proteus"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">4333</span><span class="o">:</span><span class="m">4746</span><span class="p">)</span><span class="w">
</span><span class="n">sub4cterm</span><span class="p">[[</span><span class="s2">"Rothia"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">283</span><span class="o">:</span><span class="m">702</span><span class="p">)</span><span class="w">
</span><span class="n">sub4cterm</span><span class="p">[[</span><span class="s2">"Cronobacter"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">4174</span><span class="o">:</span><span class="m">4572</span><span class="p">)</span><span class="w">
</span><span class="n">sub4cterm</span><span class="p">[[</span><span class="s2">"Prevotella"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1816</span><span class="o">:</span><span class="m">2250</span><span class="p">)</span><span class="w">
</span><span class="n">sub4all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span><span class="w">
</span><span class="n">sub4all</span><span class="p">[[</span><span class="s2">"Proteus"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'all'</span><span class="w">
</span><span class="n">sub4all</span><span class="p">[[</span><span class="s2">"Rothia"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'all'</span><span class="w">
</span><span class="n">sub4all</span><span class="p">[[</span><span class="s2">"Cronobacter"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'all'</span><span class="w">
</span><span class="n">sub4all</span><span class="p">[[</span><span class="s2">"Prevotella"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'all'</span><span class="w">

</span><span class="c1"># C terminal domains from Prevos we care about (nucleotide positions)</span><span class="w">
</span><span class="n">prevos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span><span class="w">
</span><span class="n">prevos</span><span class="p">[[</span><span class="s2">"jejuni"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1816</span><span class="o">:</span><span class="m">2250</span><span class="p">)</span><span class="w">
</span><span class="n">prevos</span><span class="p">[[</span><span class="s2">"C561"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">238</span><span class="o">:</span><span class="m">672</span><span class="p">)</span><span class="w">
</span><span class="n">prevos</span><span class="p">[[</span><span class="s2">"fusca"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">835</span><span class="o">:</span><span class="m">1257</span><span class="p">)</span><span class="w">
</span><span class="n">prevos</span><span class="p">[[</span><span class="s2">"denticola"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">3861</span><span class="o">:</span><span class="m">4295</span><span class="p">)</span><span class="w">

</span><span class="c1"># predicted Rhs start/stop sites</span><span class="w">
</span><span class="n">rhs4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span><span class="w">
</span><span class="n">rhs4</span><span class="p">[[</span><span class="s2">"Proteus"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">2329</span><span class="p">,</span><span class="m">4332</span><span class="p">)</span><span class="w">
</span><span class="n">rhs4</span><span class="p">[[</span><span class="s2">"Rothia"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">40</span><span class="p">,</span><span class="m">276</span><span class="p">)</span><span class="w">
</span><span class="n">rhs4</span><span class="p">[[</span><span class="s2">"Cronobacter"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">3937</span><span class="p">,</span><span class="m">4173</span><span class="p">)</span><span class="w">
</span><span class="n">rhs4</span><span class="p">[[</span><span class="s2">"Prevotella"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1585</span><span class="p">,</span><span class="m">1815</span><span class="p">)</span><span class="w">
</span><span class="c1"># predicted CTD start/stop sites</span><span class="w">
</span><span class="n">rhs4</span><span class="p">[[</span><span class="s2">"Proteu"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">4333</span><span class="p">,</span><span class="m">4746</span><span class="p">)</span><span class="w">
</span><span class="n">rhs4</span><span class="p">[[</span><span class="s2">"Rothi"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">283</span><span class="p">,</span><span class="m">702</span><span class="p">)</span><span class="w">
</span><span class="n">rhs4</span><span class="p">[[</span><span class="s2">"Cronobacte"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">4174</span><span class="p">,</span><span class="m">4572</span><span class="p">)</span><span class="w">
</span><span class="n">rhs4</span><span class="p">[[</span><span class="s2">"Prevotell"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1816</span><span class="p">,</span><span class="m">2250</span><span class="p">)</span><span class="w">

</span><span class="c1"># predicted region with active fold for the 4 taxa</span><span class="w">
</span><span class="n">active4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span><span class="w">
</span><span class="n">active4</span><span class="p">[[</span><span class="s2">"Proteus"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">4339</span><span class="p">,</span><span class="m">4581</span><span class="p">)</span><span class="w">
</span><span class="n">active4</span><span class="p">[[</span><span class="s2">"Rothia"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">286</span><span class="p">,</span><span class="m">546</span><span class="p">)</span><span class="w">
</span><span class="n">active4</span><span class="p">[[</span><span class="s2">"Cronobacter"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">4180</span><span class="p">,</span><span class="m">4431</span><span class="p">)</span><span class="w">
</span><span class="n">active4</span><span class="p">[[</span><span class="s2">"Prevotella"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1822</span><span class="p">,</span><span class="m">2067</span><span class="p">)</span><span class="w">

</span><span class="c1"># predicted region with active fold for the Prevotella</span><span class="w">
</span><span class="n">activePrevo</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span><span class="w">
</span><span class="n">activePrevo</span><span class="p">[[</span><span class="s2">"jejuni"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1822</span><span class="p">,</span><span class="m">2067</span><span class="p">)</span><span class="w">
</span><span class="n">activePrevo</span><span class="p">[[</span><span class="s2">"C561"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">238</span><span class="p">,</span><span class="m">492</span><span class="p">)</span><span class="w">
</span><span class="n">activePrevo</span><span class="p">[[</span><span class="s2">"fusca"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">841</span><span class="p">,</span><span class="m">1089</span><span class="p">)</span><span class="w">

</span><span class="c1"># bins/sites to focus on for plotting</span><span class="w">
</span><span class="n">subsites</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"HMP gut metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Non-HMP gut metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Human oral metagenome"</span><span class="p">)</span><span class="c1">#, "oral metagenome")</span><span class="w">
</span><span class="n">subsitesSplitHMP</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">subsites</span><span class="p">,</span><span class="w"> </span><span class="s2">"Non-HMP oral metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"HMP oral metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">subsitesWithFiji</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">subsites</span><span class="p">,</span><span class="w"> </span><span class="s2">"oral metagenome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"human metagenome"</span><span class="p">)</span><span class="w">
</span><span class="n">subsitesWithHuman</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">subsites</span><span class="p">,</span><span class="w"> </span><span class="s2">"human metagenome"</span><span class="p">)</span><span class="w">

</span><span class="n">sub4cols</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"#ee2922"</span><span class="p">,</span><span class="s2">"#68c0d3"</span><span class="p">,</span><span class="s2">"#aa9a4a"</span><span class="p">,</span><span class="s2">"#8a10aa"</span><span class="p">)</span><span class="w">
</span><span class="n">prevoCols</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"#8a10aa"</span><span class="p">,</span><span class="s2">"#ca60aa"</span><span class="p">,</span><span class="s2">"#4a109a"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>The <code class="language-plaintext highlighter-rouge">sub*</code> and <code class="language-plaintext highlighter-rouge">prevos</code> lists define the plotting windows for each taxa,
a value of <code class="language-plaintext highlighter-rouge">all</code> means all nucleotides will be used while a range of
numbers mean that only coverages or variability information in that
range will be plotted. Importantly, in the <code class="language-plaintext highlighter-rouge">plot_mean_covs_var_combo</code>
function, the filtration step happens <strong>after</strong> defining the nucleotide
range to be plotted. Thus, metagenomes are filtered based on the
proportion of non-zero-coverage positions within the nucleotide range
plotted.</p>

<p>The <code class="language-plaintext highlighter-rouge">rhs4</code> list elements with full names (e.g. <code class="language-plaintext highlighter-rouge">Proteus</code>) define the
start and stop positions for each sequence’s Rhs core domain, while the
elements with truncated names (e.g., <code class="language-plaintext highlighter-rouge">Proteu</code>; simply to add a second
element with a searchable name) define the CTD start and stop positions.</p>

<p>The <code class="language-plaintext highlighter-rouge">subsites*</code> vectors are a convenient way to command the plotter
function to focus on a certain set of metagenome categories (including
bin names).</p>

<p>With this information, we can generate the plots that become Figure
4B-D.</p>

<p>Figure 4B was generated as such:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot_mean_covs_var_combo</span><span class="p">(</span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">covs</span><span class="p">,</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">vari</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bin</span><span class="p">,</span><span class="w"> </span><span class="n">subTaxPos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sub4all</span><span class="p">,</span><span class="w"> 
                         </span><span class="n">subsite</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subsites</span><span class="p">,</span><span class="w"> </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sub4cols</span><span class="p">,</span><span class="w"> </span><span class="n">rhsBox</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rhs4</span><span class="p">,</span><span class="w"> 
                         </span><span class="n">rhsCols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s2">"#3a6b46"</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="s2">"#83ca95"</span><span class="p">,</span><span class="m">4</span><span class="p">)),</span><span class="m">2</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] "subsetting..."
## [1] "done subsetting"
## [1] "about to filter"
## [1] "done filtering"
</code></pre></div></div>

<p><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-28-1.png" alt="" /><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-28-2.png" alt="" /></p>

<p>The x-axis is nucleotide position, and each subpanel represents a
different combination of metagenome category x idrD sequence. Columns of
subpanels are different species (labelled at the top), and rows are the
metagenome categories (labelled on the left). Each line traces the
coverage provided by a single metagenome, and the number of metagenomes
for that given idrD + metagenome category combination is shown in the
upper right corner of each subpanel. The second plot generated by this
command shows the variability of each position across the different
metagenomes. The ggplot function does not like to overlay disparate y
axes, hence the separate plot. The variability is the Shannon Entropy of
all variant (and invariant) nucleotide frequencies reported by all
metagenomes for any position with variability. This is possible since
Anvi’o counted, for each metagenome, all nucleotide counts that occurred
at in a metagenome for each position with a SNP (relative to the
reference idrD sequence), which is reported in the variability file that
is <code class="language-plaintext highlighter-rouge">vari</code>.</p>

<p>It is important to note that nucleotide-level coverage patterns are
always inherently wavy, besides from real biological signal, from a
combination of stochastic/methodological, biological features
(e.g. varying GC content). For more on this, please check out <a href="http://merenlab.org/2016/12/14/coverage-variation/">this
helpful article</a>
that dives deeply into the sources of wavy coverage.</p>

<p>The variability information gives context into what is happening with
the coverage and helps us decide whether the mapping is “real”
(representing a biological population) or “spurious” (mapping artifacts
or else coverage from distant populations not considered to be part of
the population from which the reference sequence originated). For
instance, if one region has more coverage than the rest of the genome,
but also more variability, then it is a strong indicator that there is a
different population (or duplication of the gene) containing that region
but not the rest (or the rest is present in that population/duplicate
but is so divergent that bowtie2 cannot map it). Alternatively, if in
that example coverage were to go up but the variability did not, then
the same multiple-population or duplication scenario exists, but that
region is clearly more conserved at the nucleotide level between
populations/duplicates.</p>

<p>To make the in-text Figure 4B, we showed only the coverage data; for the
supplemental figure version, we overlaid the coverage figure on top of
the variability figure with <a href="https://inkscape.org/">Inkscape</a></p>

<p>Note that the FijiCOMP samples were dropped, otherwise it would look
like this:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot_mean_covs_var_combo</span><span class="p">(</span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">covs</span><span class="p">,</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">vari</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">binWithFiji</span><span class="p">,</span><span class="w"> </span><span class="n">subTaxPos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sub4all</span><span class="p">,</span><span class="w"> 
                         </span><span class="n">subsite</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subsitesWithFiji</span><span class="p">,</span><span class="w"> </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sub4cols</span><span class="p">,</span><span class="w"> </span><span class="n">rhsBox</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rhs4</span><span class="p">,</span><span class="w"> 
                         </span><span class="n">rhsCols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s2">"#3a6b46"</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="s2">"#83ca95ff"</span><span class="p">,</span><span class="m">4</span><span class="p">)),</span><span class="m">2</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] "subsetting..."
## [1] "done subsetting"
## [1] "about to filter"
## [1] "done filtering"
</code></pre></div></div>

<p><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-29-1.png" alt="" /><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-29-2.png" alt="" /></p>

<p>The FijiCOMP samples (the low-coverage “noise” added to the P. jejuni x
Human oral metagenome subpanel) are all very noisy and low-coverage, so
potentially erroneus reads or spurious mapping, so we felt safe in
discarding them.</p>

<p>We then focused in to just the C-terminal domain (CTD) - this is Figure
4C - by passing the function the nucleotide ranges in the CTD
<code class="language-plaintext highlighter-rouge">sub4cterm</code> like so:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot_mean_covs_var_combo</span><span class="p">(</span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">covs</span><span class="p">,</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">vari</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bin</span><span class="p">,</span><span class="w"> </span><span class="n">subTaxPos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sub4cterm</span><span class="p">,</span><span class="w"> 
                         </span><span class="n">subsite</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subsites</span><span class="p">,</span><span class="w"> </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sub4cols</span><span class="p">,</span><span class="w"> </span><span class="n">rhsBox</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">active4</span><span class="p">,</span><span class="w"> </span><span class="n">rhsCols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"#003380ff"</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] "subsetting..."
## [1] "done subsetting"
## [1] "about to filter"
## [1] "done filtering"
</code></pre></div></div>

<p><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-30-1.png" alt="" /><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-30-2.png" alt="" /></p>

<p>Based on this figure, we were very interested in diggng deeper into what
was going on with the <em>Prevotella</em> in the oral cavity. The enzymatic
work showed that the 5’ end of the CTD contained the active sites
(region spanned by dark blue bars) while the 3’ region was more variable
but still necessary (3’ truncations abolished activity). So, we were
curious if the precipitous drop in metagenome coverage immediately after
the first part represented a true truncation in the metagenome-sampled
population(s), or whether there truly was a 3’ region(s) in the
metagenome data but the <em>P. jejuni</em> idrD sequence was simply too
distant, in nucleotide and presumably evolutionary space, to
recruit/represent them.</p>

<p>So, we took sequences from three other <em>Prevotella</em> spp., mapped the
same human oral metagenomes against them (using the same scripts and
commands described earlier, just modified to specify the appropriate new
reference files &amp; output) - <a href="https://drive.google.com/file/d/1JVuJ5ego8dg_aP--GSca1hnEkC6T3gmA/view?usp=drive_web">coverage data
here</a>
and <a href="https://drive.google.com/file/d/1Ml6ihIBoInAAJz8-QdDCmknBhZMm9eLj/view?usp=drive_web">variability data
here</a>.</p>

<p>This produced the <code class="language-plaintext highlighter-rouge">prevo_cov</code> and <code class="language-plaintext highlighter-rouge">prevo_var</code> dataframes that have been
appearing in the scripts above. So then we plotted the coverage of their
CTDs like so:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot_mean_covs_var_combo</span><span class="p">(</span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">prevo_cov</span><span class="p">,</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">prevo_var</span><span class="p">,</span><span class="w"> </span><span class="n">proportion_gte0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bin</span><span class="p">,</span><span class="w"> 
                         </span><span class="n">subTaxPos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">prevos</span><span class="p">,</span><span class="w"> </span><span class="n">subsite</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subsites</span><span class="p">,</span><span class="w"> </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">prevoCols</span><span class="p">,</span><span class="w"> 
                         </span><span class="n">rhsBox</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">activePrevo</span><span class="p">,</span><span class="w"> </span><span class="n">rhsCols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"#003380ff"</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] "subsetting..."
## [1] "done subsetting"
## [1] "about to filter"
## [1] "done filtering"
</code></pre></div></div>

<p><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-31-1.png" alt="" /><img src="idrD_narrative_files/figure-markdown_github/unnamed-chunk-31-2.png" alt="" /></p>

<p>And it was immediately clear that the 3’ region of CTDs from <em>P.</em> sp.
C561, and <em>P. fusca</em> to a lesser extent, recruited reads, while <em>P.
jejuni</em> kept its original trend. It was also encouraging to see that the
variability went down, reflecting that having the other <em>Prevotella</em> as
alternative mapping targets ameliorated a significant amount of the
nucleotide diversity (i.e., the <em>P.</em> sp. C561 sequence was more
representative, and the sequence diversity was apparently due to other
<em>Prevotella</em> populations, not some extremely unrelated population). The
‘valley’ in the middle of the <em>P</em>. sp. C561 CTD is puzzling, and
potentially holds more insight into subpopulation diversity.</p>

        
      </section>

      <footer class="page__meta">
        
        


        
  <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2020-10-18T13:47:50-04:00">October 18, 2020</time></p>


      </footer>

      <section class="page__share">
  

  <a href="https://twitter.com/intent/tweet?text=IdrD+domain-level+metagenome+surveys%20http%3A%2F%2Flocalhost%3A4000%2Fprojects%2FIdrD" class="btn btn--twitter" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Twitter"><i class="fab fa-fw fa-twitter" aria-hidden="true"></i><span> Twitter</span></a>

  <a href="https://www.facebook.com/sharer/sharer.php?u=http%3A%2F%2Flocalhost%3A4000%2Fprojects%2FIdrD" class="btn btn--facebook" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Facebook"><i class="fab fa-fw fa-facebook" aria-hidden="true"></i><span> Facebook</span></a>

  <a href="https://www.linkedin.com/shareArticle?mini=true&url=http%3A%2F%2Flocalhost%3A4000%2Fprojects%2FIdrD" class="btn btn--linkedin" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on LinkedIn"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span> LinkedIn</span></a>
</section>


      
  <nav class="pagination">
    
      <a href="/projects/Act23_phylo" class="pagination--pager" title="Actinomyces comparative genomics
">Previous</a>
    
    
      <a href="/projects/oral_metapan" class="pagination--pager" title="Metapangenomics of Rothia and H. parainfluenzae
">Next</a>
    
  </nav>

    </div>

    
  </article>

  
  
</div>

    </div>

    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <div class="page__footer-follow">
  <ul class="social-icons">
    

    

    <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2020 Daniel Utter. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>










  </body>
</html>
