---
layout: single
title: Research focus
category: page
permalink: /Research

header:
  image: /assets/images/home_banner.svg
---
{% include JB/setup %}

I study why bacteria live where they do, what they are doing there, how they interact with each other, and how they change over time. Bacteria live everywhere we can imagine, and they often exhibit biogeography (patterns of certain bacteria appearing in certain places) across many different scales - some bacteria live only next to certain other bacteria (the micron scale), others live in certain regions of the human body but not others (centimeter scales), others might live at hydrothermal vents separated by hundreds of kilmeters.

For those who think about human health, biogeography is especially important. Many microbe-associated medical problems (think *Staph. aureus* infections) are essentially the right microbe in the wrong place. Figuring out what keeps the right microbes in the right places in healthy people is critical for medical professionals to be able to keep microbes where they're supposed to be to keep us healthy. And understanding how microbes change over time in healthy populations is essential for us to be able to distinguish between change for the better or for the worse.

# Some recent projects I've worked on:

## Thesis projects:
* How different are strains of a given bacterial species? - [[preprint](https://www.biorxiv.org/content/10.1101/2020.05.01.072496v1)]  [[methods](/projects/oral_metapan)]
  * Bacteria in the human oral microbiome exhibit striking biogeography despite salivary mixing
  * We compared the genomes of many cultured strains to compare their differences, and 
  detect patterns about where they lived, and found some exciting clues about what genes 
  underpin their biogeography
* How do oral bacterial populations change over time? - [in preparation]
  * The oral microbiome is a dynamic place, and so are bacterial populations
  * We wanted to study how bacterial populations change over short timescales, so I collected metagenomes at intervals of days to weeks from several volunteers to study microbial evolution as a dynamic process in a natural setting
* Genomics of individual-specific populations [in preparation]
  * The oral microbiome has [generally stable composition with small inter-personal variations](https://www.frontiersin.org/articles/10.3389/fmicb.2016.00564/full).
  * We are interested in studying the genetic basis of individuality to assess the ecological significance (if any) of these similar-but-distinct populations.
 	
## As collaborations:
* TM7x host range - [[article](https://www.nature.com/articles/s41396-020-00736-6)] [[methods](/projects/Act23_phylo)]
  * TM7x is an obligate bacterial symbiont that grows attached to other bacteria. Usually, TM7x is considered a parasite becuase it[impedes host growth in the lab](https://www.pnas.org/content/115/48/12277#F2).
  * Working with [Bat Bor](https://www.forsyth.org/scientists/batbileg-bor/), who found that not all hosts were affected as strongly, we compared host genomes to identify putative genes and sequence variants associated with symbiotic phenotypes.
* Toxin evolution in *Proteus mirabilis* - [[preprint](https://www.biorxiv.org/content/10.1101/2020.02.20.956912v1.full)]  [[methods](/projects/idrD_narrative)]
  * The [Gibbs lab](https://www.gibbslab.org/home), bacterial toxins used for self/non-self recognition in *Proteus mirabilis* (among other things), and part of [Dr. Denise Sirias's]() PhD research included identifying and characterizing a novel nuclease toxin. 
  * After Denise's terrific molecular work characterized the enzyme down to the domain level, we collaborated to characterize the toxin's representation in the global human metagenome and discovered how inter- and intra-species variability within metagenomes tracked domain function.